diff -ruN src55-p1/regress/lib/libc/locale/check_isw/check_isw.c src/regress/lib/libc/locale/check_isw/check_isw.c
--- src55-p1/regress/lib/libc/locale/check_isw/check_isw.c.orig	Sun Nov 10 07:31:25 2013
+++ src/regress/lib/libc/locale/check_isw/check_isw.c	Sun Nov 10 07:32:21 2013
@@ -22,11 +22,20 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <ctype.h>
+#include <locale.h>
 #include <wchar.h>
 #include <wctype.h>
 
 int bad = 0;
 
+#define p(t) printf("%s:\t ",#t); \
+	if (t) { \
+		printf("\x1b[38;5;2mOK\x1b[0m\n"); \
+	} else { \
+		bad++; \
+		printf("\x1b[38;5;1mERROR\x1b[0m\n"); \
+	}
+
 void
 check_bool(int v1, int v2, char msg)
 {
@@ -137,11 +146,30 @@
 	free(buf);
 }
 
+void
+test3()
+{
+	/** toupper, tolower single byte encoding */
+	
+	/** multi-byte encoding */
+	char *nl = setlocale(LC_ALL, "es_CO.UTF-8");
+	printf("locale %s\n", nl);
+	int c;
+	for(c = 0x41; c < 0x5A; c++) {
+		printf("c=%x ", c);
+		p(tolower(c) == c+0x20); 
+		p(toupper(c+0x20) == c); 
+		p(isalpha(c));
+		p(isalpha(c+0x20));
+	}
 
+}
+
 int
 main()
 {
 	test1();
 	test2();
+	test3();
 	return bad !=0;
 }
