diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/Makefile src/share/locale/Makefile
--- src54-s/share/locale/Makefile	Mon Oct  7 20:47:12 2013
+++ src/share/locale/Makefile	Sat Oct 12 13:23:41 2013
@@ -1,6 +1,6 @@
 #	$OpenBSD: Makefile,v 1.1 2005/08/07 10:03:44 espie Exp $
 
-SUBDIR=	cldr ctype
+SUBDIR=	cldr ctype colldef
 	
 
 .include <bsd.subdir.mk>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/Makefile src/share/locale/colldef/Makefile
--- src54-s/share/locale/colldef/Makefile	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/Makefile	Sat Oct 12 18:32:19 2013
@@ -0,0 +1,114 @@
+.include <bsd.own.mk>
+
+
+LOCALES = ISO8859-1
+LOCALES += ISO8859-15
+
+LOCALES += de.ISO8859-1
+LOCALES += de.ISO8859-15
+LOCALES_UTF8 += de
+
+LOCALES += es.ISO8859-1
+LOCALES += es.ISO8859-15
+LOCALES_UTF8 += es
+
+LOCALES += is.ISO8859-1
+LOCALES += is.ISO8859-15
+LOCALES_UTF8 += is
+
+LOCALES += no.ISO8859-1
+LOCALES += no.ISO8859-15
+LOCALES_UTF8 += no
+
+LOCALES += sv.ISO8859-1
+LOCALES += sv.ISO8859-15
+LOCALES_UTF8 += sv
+
+LOCALES += ARMSCII-8
+LOCALES += CP1251
+LOCALES += CP866
+LOCALES += ISO8859-4
+LOCALES += ISO8859-5
+LOCALES += ISO8859-7
+LOCALES += ISO8859-13
+LOCALES += KOI8-R
+LOCALES += KOI8-U
+LOCALES += ISO8859-2
+LOCALES += pl.ISO8859-2
+LOCALES += sl.ISO8859-2 
+
+LFILES += ${LOCALES:S/$/.out/g} UTF-8.out ${LOCALES_UTF8:S/$/.UTF-8.out/g}
+
+all: ${LFILES}
+
+realall: ${LFILES}
+
+.SUFFIXES: .src .out
+
+.src.out:
+	colldef -I ${.CURDIR} -o ${.TARGET} ${.IMPSRC}
+
+NOMAN=	# defined
+
+.for c in ${LOCALES}
+${c}.out: ${c}.src
+	colldef -I ${.CURDIR} -o ${c}.out ${.CURDIR}/${c}.src
+.endfor
+
+UTF-8.out: ISO8859-1.src
+	colldef -I ${.CURDIR} -o UTF-8.out ${.CURDIR}/ISO8859-1.src
+
+.for c in ${LOCALES_UTF8}
+${c}.UTF-8.out: ${c}.ISO8859-1.src
+	colldef -I ${.CURDIR} -o ${c}.UTF-8.out ${.CURDIR}/${c}.ISO8859-1.src
+.endfor
+
+
+CLEANFILES += ${LFILES}
+
+
+install:
+.for locale in ${LOCALES}
+	install -o ${LOCALEOWN} -g ${LOCALEGRP} -m ${DIRMODE} -d \
+	    ${DESTDIR}${LOCALEDIR}/${locale}
+	install -o ${LOCALEOWN} -g ${LOCALEGRP} -m ${LOCALEMODE} \
+	    ${locale}.out ${DESTDIR}${LOCALEDIR}/${locale}/LC_COLLATE
+.endfor
+	install -o ${LOCALEOWN} -g ${LOCALEGRP} -m ${DIRMODE} -d \
+	    ${DESTDIR}${LOCALEDIR}/UTF-8
+	install -o ${LOCALEOWN} -g ${LOCALEGRP} -m ${LOCALEMODE} \
+	    UTF-8.out ${DESTDIR}${LOCALEDIR}/UTF-8/LC_COLLATE
+.for locale in ${LOCALES_UTF8}
+	install -o ${LOCALEOWN} -g ${LOCALEGRP} -m ${DIRMODE} -d \
+	    ${DESTDIR}${LOCALEDIR}/${locale}.UTF-8/
+	install -o ${LOCALEOWN} -g ${LOCALEGRP} -m ${LOCALEMODE} \
+	    ${locale}.UTF-8.out ${DESTDIR}${LOCALEDIR}/${locale}.UTF-8/LC_COLLATE
+.endfor
+
+uninstall:
+.for locale in ${LOCALES}
+	if (test -f ${DESTDIR}${LOCALEDIR}/${locale}/LC_COLLATE) then { \
+	rm -f ${DESTDIR}${LOCALEDIR}/${locale}/LC_COLLATE; \
+	} fi; \
+        if (test -d ${DESTDIR}${LOCALEDIR}/${locale}/) then { \
+        r=`ls ${DESTDIR}${LOCALEDIR}/${locale}/`; \
+        if (test "$$r" = "") then { \
+        rm -rf ${DESTDIR}${LOCALEDIR}/${locale}/; \
+        } fi; \
+        } fi;
+.endfor
+	rm -f ${DESTDIR}${LOCALEDIR}/UTF-8/LC_COLLATE
+.for locale in ${LOCALES_UTF8}
+	if (test -f ${DESTDIR}${LOCALEDIR}/${locale}.UTF-8/LC_COLLATE) then { \
+	rm -f ${DESTDIR}${LOCALEDIR}/${locale}.UTF-8/LC_COLLATE; \
+	} fi; \
+        if (test -d ${DESTDIR}${LOCALEDIR}/${locale}.UTF-8/) then { \
+        r=`ls ${DESTDIR}${LOCALEDIR}/${locale}.UTF-8/`; \
+        if (test "$$r" = "") then { \
+        rm -rf ${DESTDIR}${LOCALEDIR}/${locale}.UTF-8/; \
+        } fi; \
+        } fi;
+.endfor
+
+
+.include <bsd.prog.mk>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/ARMSCII-8.src src/share/locale/colldef/ARMSCII-8.src
--- src54-s/share/locale/colldef/ARMSCII-8.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/ARMSCII-8.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,31 @@
+# $FreeBSD$
+#
+# LC_COLLATE source for Armenian/ARMSCII-8 (backward compatible with ASCII)
+# Made available by Vahe Khachikyan <vahe@khachikyan.de>
+#
+charmap map.ARMSCII-8
+substitute <armew> with "??"
+order \
+# controls
+	<armeternity>;<NU>;...;<US>;\x80;...;\xA0;\
+#
+	<SP>;(!,<armexclam>,<armaccent>);(\",<armquotleft>,<armquotright>);<Nb>;\
+	<DO>;<armemdash>;\
+	%;&;(',<armapostrophe>);(\(,<armparenleft>);(\),<armparenright>);*;+;\
+	(\,,<armcomma>);(-,<armendash>);<armyentamna>;<armellipsis>;(.,<armfullstop>);/;\
+# digits
+	0;...;9;\
+#
+	(:,<armdot>);<armsep>;\;;\<;=;>;(?,<armquestion>);<At>;\
+# capital
+	A;...;Z;\
+#
+	[;\\;];^;_;`;\
+# small
+	a;...;z;\
+# remaining punktuation etc
+	\{;|;\};~;<DT>;\
+# Armenian
+	<Armayb>;...;<armfe>;\
+# and the rest
+	\xff
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/CP1251.src src/share/locale/colldef/CP1251.src
--- src54-s/share/locale/colldef/CP1251.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/CP1251.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,43 @@
+# CP1251 (backward compatible with ASCII)
+#
+# $FreeBSD$
+#
+charmap map.CP1251
+order \
+# controls
+	<NU>;...;<US>;\
+#
+	<NS>;<SP>;!;\";<"6>;<"9>;<<<>;</>/>>;\
+	<Nb>;<Eu>;<Cu>;<DO>;\
+	%;&;';<'6>;<'9>;\(;\);*;+;<+->;\,;<-->;-;.;/;\
+# digits
+	0;...;9;\
+#
+	:;\;;\<;=;>;?;<SE>;<PI>;<Co>;<Rg>;<At>;\
+# uppercase
+	A;...;Z;\
+	<A=>;<B=>;<V=>;<G=>;<D=>;<E=>;<Z%>;<Z=>;\
+	<I=>;<J=>;<K=>;<L=>;<M=>;<N=>;<O=>;<P=>;<R=>;\
+	<S=>;<T=>;<U=>;<F=>;<H=>;<C=>;<C%>;<S%>;<Sc>;\
+	<=">;<Y=>;<%">;<JE>;<JU>;<JA>;\
+#
+	[;\\;];^;_;`;\
+# lowercase
+	a;...;z;\
+	<a=>;<b=>;<v=>;<g=>;<d=>;<e=>;<z%>;<z=>;\
+	<i=>;<j=>;<k=>;<l=>;<m=>;<n=>;<o=>;<p=>;<r=>;\
+	<s=>;<t=>;<u=>;<f=>;<h=>;<c=>;<c%>;<s%>;<sc>;\
+	<='>;<y=>;<%'>;<je>;<ju>;<ja>;\
+#
+	\{;<NO>;|;<BB>;\};~;<.M>;<DG>;<My>;\
+	<-N>;<-M>;<.9>;<:9>;</->;</=>;<.3>;<%0>;<<1>;</>1>;\
+	<N0>;<TM>;<sb>;<FB>;<DT>;\
+#
+	<D%>;<G%>;<g%>;\
+	<LJ>;<NJ>;<KJ>;<Ts>;<DZ>;\
+	<d%>;\
+	<lj>;<nj>;<kj>;<ts>;<dz>;\
+	<V%>;<v%>;<J%>;<G3>;\
+	<IO>;<IE>;<YI>;\
+	<II>;<ii>;<g3>;\
+	<io>;<ie>;<j%>;<DS>;<ds>;<yi>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/CP866.src src/share/locale/colldef/CP866.src
--- src54-s/share/locale/colldef/CP866.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/CP866.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,40 @@
+# CP866 (backward compatible with ASCII)
+#
+# $FreeBSD$
+#
+charmap map.CP866
+order \
+# controls
+	<NU>;...;<US>;\
+#
+	<NS>;<SP>;!;\";<Nb>;<Cu>;<DO>;\
+	%;&;';\(;\);*;+;\,;-;.;/;\
+# digits
+	0;...;9;\
+#
+	:;\;;\<;=;>;?;<At>;\
+# capital
+	A;...;Z;\
+	<A=>;<B=>;<V=>;<G=>;<D=>;<E=>;<IO>;<Z%>;<Z=>;\
+	<I=>;<J=>;<K=>;<L=>;<M=>;<N=>;<O=>;<P=>;<R=>;\
+	<S=>;<T=>;<U=>;<F=>;<H=>;<C=>;<C%>;<S%>;<Sc>;\
+	<=">;<Y=>;<%">;<JE>;<JU>;<JA>;\
+#
+	[;\\;];^;_;`;\
+# small
+	a;...;z;\
+	<a=>;<b=>;<v=>;<g=>;<d=>;<e=>;<io>;<z%>;<z=>;\
+	<i=>;<j=>;<k=>;<l=>;<m=>;<n=>;<o=>;<p=>;<r=>;\
+	<s=>;<t=>;<u=>;<f=>;<h=>;<c=>;<c%>;<s%>;<sc>;\
+	<='>;<y=>;<%'>;<je>;<ju>;<ja>;\
+#
+	\{;|;\};~;<.M>;<DG>;<Sb>;<N0>;<RT>;\
+#
+	<hh>;<HH>;<vv>;<VV>;<dr>;<dR>;<Dr>;<DR>;\
+	<dl>;<dL>;<Dl>;<LD>;<ur>;<uR>;<Ur>;<UR>;\
+	<ul>;<uL>;<Ul>;<UL>;<vr>;<vR>;<Vr>;<VR>;\
+	<vl>;<vL>;<Vl>;<VL>;<dh>;<dH>;<Dh>;<DH>;\
+	<uh>;<uH>;<Uh>;<UH>;<vh>;<vH>;<Vh>;<VH>;\
+	<TB>;<LB>;<FB>;<lB>;<RB>;<.S>;<:S>;<?S>;<fS>;<DT>;\
+#
+	<IE>;<ie>;<YI>;<yi>;<V%>;<v%>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/ISO8859-1.src src/share/locale/colldef/ISO8859-1.src
--- src54-s/share/locale/colldef/ISO8859-1.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/ISO8859-1.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,39 @@
+# Italian/ISO 8859-1 (backward compatible with ASCII)
+#
+# Based on de_DE.ISO8859-1
+#
+charmap map.ISO8859-1
+
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+#
+	<NS>;<SP>;!;<!I>;\";<<<>;</>/>>;<Nb>;\
+	<Cu>;<Ct>;<DO>;<Pd>;<Ye>;\
+	%;&;<',>;';\(;\);*;+;<+->;<-:>;<*X>;\,;<-->;-;.;/;\
+# digits
+	(0,<14>,<12>,<34>);(1,<1S>);(2,<2S>);(3,<3S>);4;...;9;\
+#
+	:;\;;\<;=;>;?;<?I>;<SE>;<PI>;<Co>;<Rg>;<At>;\
+# capital
+	(A,<A'>,<A!>,<A/>>,<AA>,<A:>,<A?>,<AE>);\
+	B;(C,<C,>);D;(E,<E'>,<E!>,<E/>>,<E:>);\
+	F;G;H;(I,<I'>,<I!>,<I/>>,<I:>);\
+	J;...;M;(N,<N?>);(O,<O'>,<O!>,<O/>>,<O:>,<O?>,<O//>);\
+	P;...;T;(U,<U'>,<U!>,<U/>>,<U:>);\
+	V;W;X;(Y,<Y'>);Z;\
+	<D->;<TH>;\
+#
+	[;\\;];^;<':>;_;<'m>;<''>;`;\
+# small
+	(a,<a'>,<a!>,<a/>>,<aa>,<a:>,<a?>,<ae>);\
+	b;(c,<c,>);d;(e,<e'>,<e!>,<e/>>,<e:>);\
+	f;g;h;(i,<i'>,<i!>,<i/>>,<i:>);\
+	j;...;m;(n,<n?>);(o,<o'>,<o!>,<o/>>,<o:>,<o?>,<o//>);\
+	p;...;t;(u,<u'>,<u!>,<u/>>,<u:>);\
+	v;w;x;(y,<y'>,<y:>);z;\
+	<d->;<th>;<ss>;\
+#
+	\{;<NO>;|;<BB>;\};~;<.M>;<DG>;<My>;<DT>;\
+# remains
+	<-a>;<-o>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/ISO8859-13.src src/share/locale/colldef/ISO8859-13.src
--- src54-s/share/locale/colldef/ISO8859-13.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/ISO8859-13.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,37 @@
+# Latin-7 / Baltic Rim (backward compatible with ASCII)
+#
+# Comes from la_LN.ISO8859-13 in FreeBSD
+#
+charmap map.ISO8859-13
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+# 
+	<NS>;<SP>;!;\";<"">;<Nb>;\
+	<Cu>;<DO>;<Ct>;<Pd>;\
+	%;&;';\(;\);*;+;<-:>;<*X>;<.M>;\,;<,,>;<``>;<-->;<+->;-;.;<'.>;/;\
+# digits
+	0;...;9;\
+# 
+	:;\;;\<;=;>;<<<>;</>/>>;?;<NO>;<SE>;<At>;\
+# capital
+	(A,<AA>,<A:>,<A;>,<A->,<AE>);\
+	B;(C,<C<>,<C'>);D;(E,<E'>,<E;>,<E->,<E.>);\
+	F;(G,<G,>);H;(I,<I;>,<I->);Y;\
+	J;(K,<K,>);(L,<L,>,<L//>);M;(N,<N'>,<N,>);(O,<O'>,<O:>,<O?>,<O//>,<O->);\
+	P;Q;(R,<R,>);(S,<S'>,<S<>);T;\
+	(U,<U:>,<U;>,<U->);\
+	V;W;X;(Z,<Z'>,<Z.>,<Z<>);\
+#
+	[;\\;];^;_;`;\
+# small
+	(a,<aa>,<a:>,<a;>,<a->,<ae>);\
+	b;(c,<c<>,<c'>);d;(e,<e'>,<e;>,<e->,<e.>);\
+	f;(g,<g,>);h;(i,<i;>,<i->);y;\
+	j;(k,<k,>);(l,<l,>,<l//>);m;(n,<n'>,<n,>);(o,<o'>,<o:>,<o?>,<o//>,<o->);\
+	p;q;(r,<r,>);(s,<s'>,<s<>,<ss>);t;\
+	(u,<u:>,<u;>,<u->);\
+	v;w;x;(z,<z'>,<z.>,<z<>);\
+#
+	\{;|;<BB>;\};~;<DG>;<DT>;<Co>;<Rg>;<My>;<PI>;<1S>;<2S>;<3S>;\
+	<14>;<12>;<34>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/ISO8859-15.src src/share/locale/colldef/ISO8859-15.src
--- src54-s/share/locale/colldef/ISO8859-15.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/ISO8859-15.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,39 @@
+# German/ISO 8859-1 (backward compatible with ASCII)
+#
+# $FreeBSD$
+#
+charmap map.ISO8859-15
+substitute <ss> with "ss"
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+#
+	<NS>;<SP>;!;<!I>;\";<<<>;</>/>>;<Nb>;\
+	<Eu>;<Ct>;<DO>;<Pd>;<Ye>;\
+	%;&;';\(;\);*;+;<+->;<-:>;<*X>;\,;<-->;-;.;/;\
+# digits
+	0;(1,<1S>);(2,<2S>);(3,<3S>);4;...;9;\
+#
+	:;\;;\<;=;>;?;<?I>;<SE>;<PI>;<Co>;<Rg>;<At>;\
+# capital
+	(A,<A'>,<A!>,<A/>>,<AA>,<A:>,<A?>,<AE>);\
+	B;(C,<C,>);D;(E,<E'>,<E!>,<E/>>,<E:>);\
+	F;G;H;(I,<I'>,<I!>,<I/>>,<I:>);\
+	J;...;M;(N,<N?>);(O,<O'>,<O!>,<O/>>,<O:>,<O?>,<O//>,<OE>);\
+	P;...;R;(S,<S<>);T;(U,<U'>,<U!>,<U/>>,<U:>);\
+	V;W;X;(Y,<Y'>,<Y:>);(Z,<Z<>);\
+	<D->;<TH>;\
+#
+	[;\\;];^;_;<'m>;`;\
+# small
+	(a,<a'>,<a!>,<a/>>,<aa>,<a:>,<a?>,<ae>);\
+	b;(c,<c,>);d;(e,<e'>,<e!>,<e/>>,<e:>);\
+	f;g;h;(i,<i'>,<i!>,<i/>>,<i:>);\
+	j;...;m;(n,<n?>);(o,<o'>,<o!>,<o/>>,<o:>,<o?>,<o//>,<oe>);\
+	p;...;r;(s,<s<>);t;(u,<u'>,<u!>,<u/>>,<u:>);\
+	v;w;x;(y,<y'>,<y:>);(z,<z<>);\
+	<d->;<th>;\
+#
+	\{;<NO>;|;\};~;<.M>;<DG>;<My>;<DT>;\
+# remains
+	<-a>;<-o>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/ISO8859-2.src src/share/locale/colldef/ISO8859-2.src
--- src54-s/share/locale/colldef/ISO8859-2.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/ISO8859-2.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,52 @@
+# Latin2 (for Czech; partially according to CSN 97 6030)
+# by Rudolf Cejka <cejkar@dcse.fee.vutbr.cz>
+#
+# $FreeBSD$
+#
+charmap map.ISO8859-2
+substitute <ss> with "ss"
+order \
+# spaces
+	<SP>;<NS>;\
+# capital and small
+	(A,a,<A'>,<a'>,<A/>>,<a/>>,<A(>,<a(>,<A:>,<a:>,<A;>,<a;>);\
+	(B,b);\
+	(C,c,<C'>,<c'>,<C,>,<c,>);\
+	(<C<>,<c<>);\
+	(D,d,<D<>,<d<>,<D//>,<d//>);\
+	(E,e,<E'>,<e'>,<E<>,<e<>,<E:>,<e:>,<E;>,<e;>);\
+	(F,f);\
+	(G,g);\
+	(H,h);\
+	(CH,Ch,ch);\
+	(I,i,<I'>,<i'>,<I/>>,<i/>>);\
+	(J,j);\
+	(K,k);\
+	(L,l,<L'>,<l'>,<L<>,<l<>,<L//>,<l//>);\
+	(M,m);\
+	(N,n,<N'>,<n'>,<N<>,<n<>);\
+	(O,o,<O'>,<o'>,<O/>>,<o/>>,<O:>,<o:>,<O">,<o">);\
+	(P,p);\
+	(Q,q);\
+	(R,r,<R'>,<r'>);\
+	(<R<>,<r<>);\
+	(S,s,<S'>,<s'>,<S,>,<s,>);\
+	(<S<>,<s<>);\
+	(T,t,<T<>,<t<>,<T,>,<t,>);\
+	(U,u,<U'>,<u'>,<U0>,<u0>,<U:>,<u:>,<U">,<u">);\
+	(V,v);\
+	(W,w);\
+	(X,x);\
+	(Y,y,<Y'>,<y'>);\
+	(Z,z,<Z.>,<z.>,<Z'>,<z'>);\
+	(<Z<>,<z<>);\
+# digits
+	0;...;9;\
+# symbols
+	.;\,;\;;?;!;:;';`;\";\
+	<-->;-;|;/;\\;\(;\);[;];\<;>;\{;\};\
+	&;<SE>;%;<DO>;<At>;\
+	_;=;^;<-:>;+;<*X>;*;<Nb>;~;<DG>;<Cu>;\
+	<'.>;<''>;<'<>;<'(>;<':>;<'">;<',>;<';>;\
+# controls
+	<NU>;...;<US>;<DT>;<PA>;...;<AC>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/ISO8859-4.src src/share/locale/colldef/ISO8859-4.src
--- src54-s/share/locale/colldef/ISO8859-4.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/ISO8859-4.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,36 @@
+# Lithuanian (backward compatible with ASCII)
+#
+# $FreeBSD$
+#
+charmap map.ISO8859-4
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+# 
+	<NS>;<SP>;!;\";<Nb>;\
+	<Cu>;<DO>;\
+	%;&;<',>;<';>;';\(;\);*;+;<-:>;<*X>;\,;<-->;-;.;<'.>;/;\
+# digits
+	0;...;9;\
+# 
+	:;\;;\<;=;>;?;<SE>;<At>;\
+# capital
+	(A,<A'>,<A/>>,<AA>,<A:>,<A;>,<A?>,<A->,<AE>);\
+	B;(C,<C<>);(D,<D//>);(E,<E'>,<E:>,<E;>,<E->,<E.>);\
+	F;(G,<G,>);H;(I,<I'>,<I/>>,<I?>,<I;>,<I->);Y;\
+	J;(K,<K,>);(L,<L,>);M;(N,<N,>);<NG>;(O,<O/>>,<O:>,<O?>,<O//>,<O->);\
+	P;Q;(R,<R,>);(S,<S<>);(T,<T//>);\
+	(U,<U'>,<U/>>,<U:>,<U;>,<U?>,<U->);\
+	V;W;X;(Z,<Z<>);\
+#
+	[;\\;];^;<':>;<'<>;_;<'m>;<''>;`;\
+# small
+	(a,<a'>,<a/>>,<aa>,<a:>,<a?>,<a;>,<a->,<ae>);\
+	b;(c,<c<>);(d,<d//>);(e,<e'>,<e:>,<e;>,<e->,<e.>);\
+	f;(g,<g,>);h;(i,<i'>,<i/>>,<i?>,<i;>,<i->);y;\
+	j;(k,<k,>);<kk>;(l,<l,>);m;(n,<n,>);<ng>;(o,<o/>>,<o:>,<o?>,<o//>,<o->);\
+	p;q;(r,<r,>);(s,<s<>,<ss>);(t,<t//>);\
+	(u,<u'>,<u/>>,<u:>,<u;>,<u?>,<u->);\
+	v;w;x;(z,<z<>);\
+#
+	\{;|;\};~;<DG>;<DT>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/ISO8859-5.src src/share/locale/colldef/ISO8859-5.src
--- src54-s/share/locale/colldef/ISO8859-5.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/ISO8859-5.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,39 @@
+# Russian part of ISO8859-5 (backward compatible with ASCII)
+#
+# $FreeBSD$
+#
+charmap map.ISO8859-5
+order \
+# controls
+	<NU>;...;<US>;\
+#
+	<NS>;<SP>;!;\";<Nb>;<DO>;\
+	%;&;';\(;\);*;+;\,;<-->;-;.;/;\
+# digits
+	0;...;9;\
+#
+	:;\;;\<;=;>;?;<At>;\
+# capital
+	A;...;Z;\
+	<A=>;<B=>;<V=>;<G=>;<D=>;<E=>;<IO>;<Z%>;<Z=>;\
+	<I=>;<J=>;<K=>;<L=>;<M=>;<N=>;<O=>;<P=>;<R=>;\
+	<S=>;<T=>;<U=>;<F=>;<H=>;<C=>;<C%>;<S%>;<Sc>;\
+	<=">;<Y=>;<%">;<JE>;<JU>;<JA>;\
+#
+	[;\\;];^;_;`;\
+# small
+	a;...;z;\
+	<a=>;<b=>;<v=>;<g=>;<d=>;<e=>;<io>;<z%>;<z=>;\
+	<i=>;<j=>;<k=>;<l=>;<m=>;<n=>;<o=>;<p=>;<r=>;\
+	<s=>;<t=>;<u=>;<f=>;<h=>;<c=>;<c%>;<s%>;<sc>;\
+	<='>;<y=>;<%'>;<je>;<ju>;<ja>;\
+#
+	\{;|;\};~;<N0>;<DT>;\
+#
+	<PA>;<HO>;<BH>;<NH>;<IN>;<NL>;\
+	<SA>;<ES>;<HS>;<HJ>;<VS>;<PD>;<PU>;<RI>;<S2>;\
+	<S3>;<DC>;<P1>;<P2>;<TS>;<CC>;<MW>;<SG>;<EG>;\
+	<SS>;<GC>;<SC>;<CI>;<ST>;<OC>;<PM>;<AC>;<D%>;\
+	<G%>;<IE>;<DS>;<II>;<YI>;<J%>;<LJ>;<NJ>;<Ts>;\
+	<KJ>;<V%>;<DZ>;<d%>;<g%>;<ie>;<ds>;\
+	<ii>;<yi>;<j%>;<lj>;<nj>;<ts>;<kj>;<SE>;<v%>;<dz>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/ISO8859-7.src src/share/locale/colldef/ISO8859-7.src
--- src54-s/share/locale/colldef/ISO8859-7.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/ISO8859-7.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,34 @@
+# $FreeBSD$
+#
+# Greek/ISO 8859-7 (backward compatible with ASCII)
+#
+charmap map.ISO8859-7
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+#
+	<NS>;<SP>;!;\";<<<>;</>/>>;<Nb>;<DO>;<Pd>;%;&;\
+	';\(;\);*;+;<+->;\,;<-->;-;.;/;\
+# digits
+	(0,<12>);1;(2,<2S>);(3,<3S>);4;...;9;\
+#
+	:;\;;\<;=;>;?;<SE>;<Co>;<At>;\
+# capital
+	A;...;Z;\
+	(<A%>,<A*>);<B*>;<G*>;<D*>;(<E%>,<E*>);<Z*>;(<Y%>,\
+	<Y*>);<H*>;(<J*>,<I%>,<I*>);<K*>;<L*>;<M*>;<N*>;\
+	<C*>;(<O%>,<O*>);<P*>;<R*>;<S*>;<T*>;(<V*>,<U%>,\
+	<U*>);<F*>;<X*>;<Q*>;(<W%>,<W*>);\
+# small
+	a;...;z;\
+	(<a%>,<a*>);<b*>;<g*>;<d*>;(<e%>,<e*>);<z*>;(<y%>,\
+	<y*>);<h*>;(<i3>,<j*>,<i%>,<i*>);<k*>;<l*>;<m*>;\
+	<n*>;<c*>;(<o%>,<o*>);<p*>;<r*>;<s*>;<*s>;<t*>;\
+	(<u3>,<v*>,<u%>,<u*>);<f*>;<x*>;<q*>;(<w%>,<w*>);\
+#
+	[;\\;];<':>;_;<''>;`;\
+#
+	\{;<NO>;|;<BB>;\};~;<.M>;<DG>;<DT>;\
+# remains
+	<'/>>;<9'>;<'9>;<IE>;<DS>;<NJ>;<V%>;<-M>;<'%>;\
+	<v=>;<dz>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/KOI8-R.src src/share/locale/colldef/KOI8-R.src
--- src54-s/share/locale/colldef/KOI8-R.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/KOI8-R.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,39 @@
+# koi8-r (backward compatible with ASCII)
+#
+# $FreeBSD$
+#
+charmap map.KOI8-R
+order \
+# controls
+	<NU>;...;<US>;\
+#
+	<NS>;<SP>;!;\";<Nb>;<DO>;\
+	%;&;';\(;\);*;+;<-:>;\,;-;.;/;\
+# digits
+	0;1;(2,<2S>);3;...;9;\
+#
+	:;\;;\<;<=<>;=;</>=>;>;?;<Co>;<At>;\
+# capital
+	A;...;Z;\
+	<A=>;<B=>;<V=>;<G=>;<D=>;<E=>;<IO>;<Z%>;<Z=>;\
+	<I=>;<J=>;<K=>;<L=>;<M=>;<N=>;<O=>;<P=>;<R=>;\
+	<S=>;<T=>;<U=>;<F=>;<H=>;<C=>;<C%>;<S%>;<Sc>;\
+	<=">;<Y=>;<%">;<JE>;<JU>;<JA>;\
+#
+	[;\\;];^;_;`;\
+# small
+	a;...;z;\
+	<a=>;<b=>;<v=>;<g=>;<d=>;<e=>;<io>;<z%>;<z=>;\
+	<i=>;<j=>;<k=>;<l=>;<m=>;<n=>;<o=>;<p=>;<r=>;\
+	<s=>;<t=>;<u=>;<f=>;<h=>;<c=>;<c%>;<s%>;<sc>;\
+	<='>;<y=>;<%'>;<je>;<ju>;<ja>;\
+#
+	\{;|;\};~;<.M>;<DG>;\
+#
+	<Sb>;<RT>;<?2>;<Iu>;<Il>;\
+	<hh>;<HH>;<vv>;<VV>;<dr>;<dR>;<Dr>;<DR>;\
+	<dl>;<dL>;<Dl>;<LD>;<ur>;<uR>;<Ur>;<UR>;\
+	<ul>;<uL>;<Ul>;<UL>;<vr>;<vR>;<Vr>;<VR>;\
+	<vl>;<vL>;<Vl>;<VL>;<dh>;<dH>;<Dh>;<DH>;\
+	<uh>;<uH>;<Uh>;<UH>;<vh>;<vH>;<Vh>;<VH>;\
+	<TB>;<LB>;<FB>;<lB>;<RB>;<.S>;<:S>;<?S>;<fS>;<DT>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/KOI8-U.src src/share/locale/colldef/KOI8-U.src
--- src54-s/share/locale/colldef/KOI8-U.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/KOI8-U.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,39 @@
+# koi8-u (backward compatible with ASCII)
+#
+# $FreeBSD$
+#
+charmap map.KOI8-U
+order \
+# controls
+	<NU>;...;<US>;\
+#
+	<NS>;<SP>;!;\";<Nb>;<DO>;\
+	%;&;';\(;\);*;+;<-:>;\,;-;.;/;\
+# digits
+	0;1;(2,<2S>);3;...;9;\
+#
+	:;\;;\<;<=<>;=;</>=>;>;?;<Co>;<At>;\
+# capital
+	A;...;Z;\
+	<A=>;<B=>;<V=>;<G=>;<G3>;<D=>;<E=>;<IE>;<IO>;<Z%>;<Z=>;\
+	<I=>;<II>;<YI>;<J=>;<K=>;<L=>;<M=>;<N=>;<O=>;<P=>;<R=>;\
+	<S=>;<T=>;<U=>;<F=>;<H=>;<C=>;<C%>;<S%>;<Sc>;\
+	<=">;<Y=>;<JE>;<JU>;<JA>;<%">;\
+#
+	[;\\;];^;_;`;\
+# small
+	a;...;z;\
+	<a=>;<b=>;<v=>;<g=>;<g3>;<d=>;<e=>;<ie>;<io>;<z%>;<z=>;\
+	<i=>;<ii>;<yi>;<j=>;<k=>;<l=>;<m=>;<n=>;<o=>;<p=>;<r=>;\
+	<s=>;<t=>;<u=>;<f=>;<h=>;<c=>;<c%>;<s%>;<sc>;\
+	<='>;<y=>;<je>;<ju>;<ja>;<%'>;\
+#
+	\{;|;\};~;<.M>;<DG>;\
+#
+	<Sb>;<RT>;<?2>;<Iu>;<Il>;\
+	<hh>;<HH>;<vv>;<VV>;<dr>;<dR>;<DR>;\
+	<dl>;<LD>;<ur>;<uR>;<Ur>;<UR>;\
+	<ul>;<uL>;<UL>;<vr>;<vR>;<Vr>;<VR>;\
+	<vl>;<vL>;<VL>;<dh>;<DH>;\
+	<uh>;<uH>;<Uh>;<UH>;<vh>;<vH>;<VH>;\
+	<TB>;<LB>;<FB>;<lB>;<RB>;<.S>;<:S>;<?S>;<fS>;<DT>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/de.ISO8859-1.src src/share/locale/colldef/de.ISO8859-1.src
--- src54-s/share/locale/colldef/de.ISO8859-1.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/de.ISO8859-1.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,39 @@
+# German/ISO 8859-1 (backward compatible with ASCII)
+#
+# $FreeBSD$
+#
+charmap map.ISO8859-1
+substitute <ss> with "ss"
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+#
+	<NS>;<SP>;!;<!I>;\";<<<>;</>/>>;<Nb>;\
+	<Cu>;<Ct>;<DO>;<Pd>;<Ye>;\
+	%;&;<',>;';\(;\);*;+;<+->;<-:>;<*X>;\,;<-->;-;.;/;\
+# digits
+	(0,<14>,<12>,<34>);(1,<1S>);(2,<2S>);(3,<3S>);4;...;9;\
+#
+	:;\;;\<;=;>;?;<?I>;<SE>;<PI>;<Co>;<Rg>;<At>;\
+# capital
+	(A,<A'>,<A!>,<A/>>,<AA>,<A:>,<A?>,<AE>);\
+	B;(C,<C,>);D;(E,<E'>,<E!>,<E/>>,<E:>);\
+	F;G;H;(I,<I'>,<I!>,<I/>>,<I:>);\
+	J;...;M;(N,<N?>);(O,<O'>,<O!>,<O/>>,<O:>,<O?>,<O//>);\
+	P;...;T;(U,<U'>,<U!>,<U/>>,<U:>);\
+	V;W;X;(Y,<Y'>);Z;\
+	<D->;<TH>;\
+#
+	[;\\;];^;<':>;_;<'m>;<''>;`;\
+# small
+	(a,<a'>,<a!>,<a/>>,<aa>,<a:>,<a?>,<ae>);\
+	b;(c,<c,>);d;(e,<e'>,<e!>,<e/>>,<e:>);\
+	f;g;h;(i,<i'>,<i!>,<i/>>,<i:>);\
+	j;...;m;(n,<n?>);(o,<o'>,<o!>,<o/>>,<o:>,<o?>,<o//>);\
+	p;...;t;(u,<u'>,<u!>,<u/>>,<u:>);\
+	v;w;x;(y,<y'>,<y:>);z;\
+	<d->;<th>;\
+#
+	\{;<NO>;|;<BB>;\};~;<.M>;<DG>;<My>;<DT>;\
+# remains
+	<-a>;<-o>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/de.ISO8859-15.src src/share/locale/colldef/de.ISO8859-15.src
--- src54-s/share/locale/colldef/de.ISO8859-15.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/de.ISO8859-15.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,39 @@
+# German/ISO 8859-1 (backward compatible with ASCII)
+#
+# $FreeBSD$
+#
+charmap map.ISO8859-15
+substitute <ss> with "ss"
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+#
+	<NS>;<SP>;!;<!I>;\";<<<>;</>/>>;<Nb>;\
+	<Eu>;<Ct>;<DO>;<Pd>;<Ye>;\
+	%;&;';\(;\);*;+;<+->;<-:>;<*X>;\,;<-->;-;.;/;\
+# digits
+	0;(1,<1S>);(2,<2S>);(3,<3S>);4;...;9;\
+#
+	:;\;;\<;=;>;?;<?I>;<SE>;<PI>;<Co>;<Rg>;<At>;\
+# capital
+	(A,<A'>,<A!>,<A/>>,<AA>,<A:>,<A?>,<AE>);\
+	B;(C,<C,>);D;(E,<E'>,<E!>,<E/>>,<E:>);\
+	F;G;H;(I,<I'>,<I!>,<I/>>,<I:>);\
+	J;...;M;(N,<N?>);(O,<O'>,<O!>,<O/>>,<O:>,<O?>,<O//>,<OE>);\
+	P;...;R;(S,<S<>);T;(U,<U'>,<U!>,<U/>>,<U:>);\
+	V;W;X;(Y,<Y'>,<Y:>);(Z,<Z<>);\
+	<D->;<TH>;\
+#
+	[;\\;];^;_;<'m>;`;\
+# small
+	(a,<a'>,<a!>,<a/>>,<aa>,<a:>,<a?>,<ae>);\
+	b;(c,<c,>);d;(e,<e'>,<e!>,<e/>>,<e:>);\
+	f;g;h;(i,<i'>,<i!>,<i/>>,<i:>);\
+	j;...;m;(n,<n?>);(o,<o'>,<o!>,<o/>>,<o:>,<o?>,<o//>,<oe>);\
+	p;...;r;(s,<s<>);t;(u,<u'>,<u!>,<u/>>,<u:>);\
+	v;w;x;(y,<y'>,<y:>);(z,<z<>);\
+	<d->;<th>;\
+#
+	\{;<NO>;|;\};~;<.M>;<DG>;<My>;<DT>;\
+# remains
+	<-a>;<-o>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/es.ISO8859-1.src src/share/locale/colldef/es.ISO8859-1.src
--- src54-s/share/locale/colldef/es.ISO8859-1.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/es.ISO8859-1.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,38 @@
+# Espan~ol (backward compatible with ASCII)
+#
+# Adapted and fixed from FreeBSD (es_ES.ISO8859-1) 
+#
+charmap map.ISO8859-1
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+#
+	<NS>;<SP>;!;<!I>;\";<<<>;</>/>>;<Nb>;\
+	<Cu>;<Ct>;<DO>;<Pd>;<Ye>;\
+	%;&;<',>;';\(;\);*;+;<+->;<-:>;<*X>;\,;<-->;-;.;/;\
+# digits
+	(0,<14>,<12>,<34>);(1,<1S>);(2,<2S>);(3,<3S>);4;...;9;\
+#
+	:;\;;\<;=;>;?;<?I>;<SE>;<PI>;<Co>;<Rg>;<At>;\
+# capital
+	(A,<A'>,<A!>,<A/>>,<AA>,<A:>,<A?>,<AE>);\
+	B;(C,<C,>);{CH,Ch};D;(E,<E'>,<E!>,<E/>>,<E:>);\
+	F;G;H;(I,<I'>,<I!>,<I/>>,<I:>);\
+	J;...;K;L;{LL,Ll};M;N;<N?>;(O,<O'>,<O!>,<O/>>,<O:>,<O?>,<O//>);\
+	P;Q;R;{RR,Rr};S;T;(U,<U'>,<U!>,<U/>>,<U:>);\
+	V;W;X;(Y,<Y'>);Z;\
+	<D->;<TH>;\
+#
+	[;\\;];^;<':>;_;<'m>;<''>;`;\
+# small
+	(a,<a'>,<a!>,<a/>>,<aa>,<a:>,<a?>,<ae>);\
+	b;(c,<c,>);ch;d;(e,<e'>,<e!>,<e/>>,<e:>);\
+	f;g;h;(i,<i'>,<i!>,<i/>>,<i:>);\
+	j;...;k;l;ll;m;n;<n?>;(o,<o'>,<o!>,<o/>>,<o:>,<o?>,<o//>);\
+	p;q;r;rr;s;t;(u,<u'>,<u!>,<u/>>,<u:>);\
+	v;w;x;(y,<y'>,<y:>);z;\
+	<d->;<th>;<ss>;\
+#
+	\{;<NO>;|;<BB>;\};~;<.M>;<DG>;<My>;<DT>;\
+# remains
+	<-a>;<-o>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/es.ISO8859-15.src src/share/locale/colldef/es.ISO8859-15.src
--- src54-s/share/locale/colldef/es.ISO8859-15.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/es.ISO8859-15.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,38 @@
+# Espan~ol (backward compatible with ASCII)
+#
+# Copied from FreeBSD es_ES.ISO8859-15 and fixed
+#
+charmap map.ISO8859-15
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+#
+	<NS>;<SP>;!;<!I>;\";<<<>;</>/>>;<Nb>;\
+	<Eu>;<Ct>;<DO>;<Pd>;<Ye>;\
+	%;&;';\(;\);*;+;<+->;<-:>;<*X>;\,;<-->;-;.;/;\
+# digits
+	0;(1,<1S>);(2,<2S>);(3,<3S>);4;...;9;\
+#
+	:;\;;\<;=;>;?;<?I>;<SE>;<PI>;<Co>;<Rg>;<At>;\
+# capital
+	(A,<A'>,<A!>,<A/>>,<AA>,<A:>,<A?>,<AE>);\
+	B;(C,<C,>);{CH,Ch};D;(E,<E'>,<E!>,<E/>>,<E:>);\
+	F;G;H;(I,<I'>,<I!>,<I/>>,<I:>);\
+	J;...;K;L;{LL,Ll};M;N;<N?>;(O,<O'>,<O!>,<O/>>,<O:>,<O?>,<O//>,<OE>);\
+	P;Q;R;{RR,Rr};(S,<S<>);T;(U,<U'>,<U!>,<U/>>,<U:>);\
+	V;W;X;(Y,<Y'>,<Y:>);(Z,<Z<>);\
+	<D->;<TH>;\
+#
+	[;\\;];^;_;<'m>;`;\
+# small
+	(a,<a'>,<a!>,<a/>>,<aa>,<a:>,<a?>,<ae>);\
+	b;(c,<c,>);ch;d;(e,<e'>,<e!>,<e/>>,<e:>);\
+	f;g;h;(i,<i'>,<i!>,<i/>>,<i:>);\
+	j;...;k;l;ll;m;n;<n?>;(o,<o'>,<o!>,<o/>>,<o:>,<o?>,<o//>,<oe>);\
+	p;q;r;rr;(s,<s<>);t;(u,<u'>,<u!>,<u/>>,<u:>);\
+	v;w;x;(y,<y'>,<y:>);(z,<z<>);\
+	<d->;<th>;<ss>;\
+#
+	\{;<NO>;|;\};~;<.M>;<DG>;<My>;<DT>;\
+# remains
+	<-a>;<-o>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/is.ISO8859-1.src src/share/locale/colldef/is.ISO8859-1.src
--- src54-s/share/locale/colldef/is.ISO8859-1.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/is.ISO8859-1.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,38 @@
+# icelandic (backward compatible with ASCII)
+#
+# $FreeBSD$
+#
+charmap map.ISO8859-1
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+#
+	<NS>;<SP>;!;<!I>;\";<<<>;</>/>>;<Nb>;\
+	<Cu>;<Ct>;<DO>;<Pd>;<Ye>;\
+	%;&;<',>;';\(;\);*;+;<+->;<-:>;<*X>;\,;<-->;-;.;/;\
+# digits
+	(0,<14>,<12>,<34>);(1,<1S>);(2,<2S>);(3,<3S>);4;...;9;\
+#
+	:;\;;\<;=;>;?;<?I>;<SE>;<PI>;<Co>;<Rg>;<At>;\
+# capital
+	(A,<A'>,<A!>,<A/>>,<AA>,<A:>,<A?>);\
+	B;(C,<C,>);(D,<D->);(E,<E'>,<E!>,<E/>>,<E:>);\
+	F;G;H;(I,<I'>,<I!>,<I/>>,<I:>);\
+	J;...;M;(N,<N?>);(O,<O'>,<O!>,<O/>>,<O?>,<O//>);\
+	P;...;T;(U,<U'>,<U!>,<U/>>,<U:>);\
+	V;W;X;(Y,<Y'>);Z;\
+	<TH>;<AE>;<O:>;\
+#
+	[;\\;];^;<':>;_;<'m>;<''>;`;\
+# small
+	(a,<a'>,<a!>,<a/>>,<aa>,<a:>,<a?>);\
+	b;(c,<c,>);(d,<d->);(e,<e'>,<e!>,<e/>>,<e:>);\
+	f;g;h;(i,<i'>,<i!>,<i/>>,<i:>);\
+	j;...;m;(n,<n?>);(o,<o'>,<o!>,<o/>>,<o?>,<o//>);\
+	p;...;t;(u,<u'>,<u!>,<u/>>,<u:>);\
+	v;w;x;(y,<y'>,<y:>);z;\
+	<th>;<ae>;<o:>;<ss>;\
+#
+	\{;<NO>;|;<BB>;\};~;<.M>;<DG>;<My>;<DT>;\
+# remains
+	<-a>;<-o>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/is.ISO8859-15.src src/share/locale/colldef/is.ISO8859-15.src
--- src54-s/share/locale/colldef/is.ISO8859-15.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/is.ISO8859-15.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,38 @@
+# icelandic (backward compatible with ASCII)
+#
+# $FreeBSD$
+#
+charmap map.ISO8859-15
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+#
+	<NS>;<SP>;!;<!I>;\";<<<>;</>/>>;<Nb>;\
+	<Eu>;<Ct>;<DO>;<Pd>;<Ye>;\
+	%;&;';\(;\);*;+;<+->;<-:>;<*X>;\,;<-->;-;.;/;\
+# digits
+	0;(1,<1S>);(2,<2S>);(3,<3S>);4;...;9;\
+#
+	:;\;;\<;=;>;?;<?I>;<SE>;<PI>;<Co>;<Rg>;<At>;\
+# capital
+	(A,<A'>,<A!>,<A/>>,<AA>,<A:>,<A?>);\
+	B;(C,<C,>);(D,<D->);(E,<E'>,<E!>,<E/>>,<E:>);\
+	F;G;H;(I,<I'>,<I!>,<I/>>,<I:>);\
+	J;...;M;(N,<N?>);(O,<O'>,<O!>,<O/>>,<O?>,<O//>,<OE>);\
+	P;...;R;(S,<S<>);T;(U,<U'>,<U!>,<U/>>,<U:>);\
+	V;W;X;(Y,<Y'>,<Y:>);(Z,<Z<>);\
+	<TH>;<AE>;<O:>;\
+#
+	[;\\;];^;_;<'m>;`;\
+# small
+	(a,<a'>,<a!>,<a/>>,<aa>,<a:>,<a?>);\
+	b;(c,<c,>);(d,<d->);(e,<e'>,<e!>,<e/>>,<e:>);\
+	f;g;h;(i,<i'>,<i!>,<i/>>,<i:>);\
+	j;...;m;(n,<n?>);(o,<o'>,<o!>,<o/>>,<o?>,<o//>,<oe>);\
+	p;...;r;(s,<s<>);t;(u,<u'>,<u!>,<u/>>,<u:>);\
+	v;w;x;(y,<y'>,<y:>);(z,<z<>);\
+	<th>;<ae>;<o:>;<ss>;\
+#
+	\{;<NO>;|;\};~;<.M>;<DG>;<My>;<DT>;\
+# remains
+	<-a>;<-o>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/map.ARMSCII-8 src/share/locale/colldef/map.ARMSCII-8
--- src54-s/share/locale/colldef/map.ARMSCII-8	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/map.ARMSCII-8	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,144 @@
+# $FreeBSD$
+#
+# map.ARMSCII-8 defined by Vahe Khachikyan <vahe@khachikyan.de>
+# 
+NU \x00
+SH \x01
+SX \x02
+EX \x03
+ET \x04
+EQ \x05
+AK \x06
+BL \x07
+BS \x08
+HT \x09
+LF \x0a
+VT \x0b
+FF \x0c
+CR \x0d
+SO \x0e
+SI \x0f
+DL \x10
+D1 \x11
+D2 \x12
+D3 \x13
+D4 \x14
+NK \x15
+SY \x16
+EB \x17
+CN \x18
+EM \x19
+SB \x1a
+EC \x1b
+FS \x1c
+GS \x1d
+RS \x1e
+US \x1f
+SP \x20
+Nb \x23
+DO \x24
+At \x40
+<( \x5b
+// \x5c
+)> \x5d
+'> \x5e
+'! \x60
+(! \x7b
+!! \x7c
+!) \x7d
+'? \x7e
+DT \x7f
+armeternity	\xA1
+armew		\xA2
+armfullstop	\xA3
+armparenright	\xA4
+armparenleft	\xA5
+armquotright	\xA6
+armquotleft	\xA7
+armemdash	\xA8
+armdot		\xA9
+armsep		\xAA
+armcomma	\xAB
+armendash	\xAC
+armyentamna	\xAD
+armellipsis	\xAE
+armexclam	\xAF
+armaccent	\xB0
+armquestion	\xB1
+Armayb		\xB2
+armayb		\xB3
+Armben		\xB4
+armben		\xB5
+Armgim		\xB6
+armgim		\xB7
+Armda		\xB8
+armda		\xB9
+Armyech		\xBA
+armyech		\xBB
+Armza		\xBC
+armza		\xBD
+Arme		\xBE
+arme		\xBF
+Armat		\xC0
+armat		\xC1
+Armto		\xC2
+armto		\xC3
+Armzhe		\xC4
+armzhe		\xC5
+Armini		\xC6
+armini		\xC7
+Armlyun		\xC8
+armlyun		\xC9
+Armkhe		\xCA
+armkhe		\xCB
+Armtsa		\xCC
+armtsa		\xCD
+Armken		\xCE
+armken		\xCF
+Armho		\xD0
+armho		\xD1
+Armdza		\xD2
+armdza		\xD3
+Armghat		\xD4
+armghat		\xD5
+Armtche		\xD6
+armtche		\xD7
+Armmen		\xD8
+armmen		\xD9
+Armhi		\xDA
+armhi		\xDB
+Armnu		\xDC
+armnu		\xDD
+Armsha		\xDE
+armsha		\xDF
+Armvo		\xE0
+armvo		\xE1
+Armcha		\xE2
+armcha		\xE3
+Armpe		\xE4
+armpe		\xE5
+Armje		\xE6
+armje		\xE7
+Armra		\xE8
+armra		\xE9
+Armse		\xEA
+armse		\xEB
+Armvev		\xEC
+armvev		\xED
+Armtyun		\xEE
+armtyun		\xEF
+Armre		\xF0
+armre		\xF1
+Armtso		\xF2
+armtso		\xF3
+Armvyun		\xF4
+armvyun		\xF5
+Armpyur		\xF6
+armpyur		\xF7
+Armke		\xF8
+armke		\xF9
+Armo		\xFA
+armo		\xFB
+Armfe		\xFC
+armfe		\xFD
+armapostrophe	\xFE
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/map.CP1251 src/share/locale/colldef/map.CP1251
--- src54-s/share/locale/colldef/map.CP1251	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/map.CP1251	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,257 @@
+# $FreeBSD$
+NU \x00
+SH \x01
+SX \x02
+EX \x03
+ET \x04
+EQ \x05
+AK \x06
+BL \x07
+BS \x08
+HT \x09
+LF \x0a
+VT \x0b
+FF \x0c
+CR \x0d
+SO \x0e
+SI \x0f
+DL \x10
+D1 \x11
+D2 \x12
+D3 \x13
+D4 \x14
+NK \x15
+SY \x16
+EB \x17
+CN \x18
+EM \x19
+SB \x1a
+EC \x1b
+FS \x1c
+GS \x1d
+RS \x1e
+US \x1f
+SP \x20
+! \x21
+" \x22
+Nb \x23
+DO \x24
+% \x25
+& \x26
+' \x27
+( \x28
+) \x29
+* \x2a
++ \x2b
+, \x2c
+- \x2d
+. \x2e
+/ \x2f
+0 \x30
+1 \x31
+2 \x32
+3 \x33
+4 \x34
+5 \x35
+6 \x36
+7 \x37
+8 \x38
+9 \x39
+: \x3a
+; \x3b
+< \x3c
+= \x3d
+> \x3e
+? \x3f
+At \x40
+A \x41
+B \x42
+C \x43
+D \x44
+E \x45
+F \x46
+G \x47
+H \x48
+I \x49
+J \x4a
+K \x4b
+L \x4c
+M \x4d
+N \x4e
+O \x4f
+P \x50
+Q \x51
+R \x52
+S \x53
+T \x54
+U \x55
+V \x56
+W \x57
+X \x58
+Y \x59
+Z \x5a
+<( \x5b
+// \x5c
+)> \x5d
+'> \x5e
+_ \x5f
+'! \x60
+a \x61
+b \x62
+c \x63
+d \x64
+e \x65
+f \x66
+g \x67
+h \x68
+i \x69
+j \x6a
+k \x6b
+l \x6c
+m \x6d
+n \x6e
+o \x6f
+p \x70
+q \x71
+r \x72
+s \x73
+t \x74
+u \x75
+v \x76
+w \x77
+x \x78
+y \x79
+z \x7a
+(! \x7b
+!! \x7c
+!) \x7d
+'? \x7e
+DT \x7f
+D% \x80
+G% \x81
+.9 \x82
+g% \x83
+:9 \x84
+.3 \x85
+/- \x86
+/= \x87
+Eu \x88
+%0 \x89
+LJ \x8a
+<1 \x8b
+NJ \x8c
+KJ \x8d
+Ts \x8e
+DZ \x8f
+d% \x90
+'6 \x91
+'9 \x92
+"6 \x93
+"9 \x94
+sb \x95
+-N \x96
+-M \x97
+FB \x98
+TM \x99
+lj \x9a
+>1 \x9b
+nj \x9c
+kj \x9d
+ts \x9e
+dz \x9f
+NS \xa0
+V% \xa1
+v% \xa2
+J% \xa3
+Cu \xa4
+G3 \xa5
+BB \xa6
+SE \xa7
+IO \xa8
+Co \xa9
+IE \xaa
+<< \xab
+NO \xac
+-- \xad
+Rg \xae
+YI \xaf
+DG \xb0
++- \xb1
+II \xb2
+ii \xb3
+g3 \xb4
+My \xb5
+PI \xb6
+.M \xb7
+io \xb8
+N0 \xb9
+ie \xba
+>> \xbb
+j% \xbc
+DS \xbd
+ds \xbe
+yi \xbf
+A= \xc0
+B= \xc1
+V= \xc2
+G= \xc3
+D= \xc4
+E= \xc5
+Z% \xc6
+Z= \xc7
+I= \xc8
+J= \xc9
+K= \xca
+L= \xcb
+M= \xcc
+N= \xcd
+O= \xce
+P= \xcf
+R= \xd0
+S= \xd1
+T= \xd2
+U= \xd3
+F= \xd4
+H= \xd5
+C= \xd6
+C% \xd7
+S% \xd8
+Sc \xd9
+=" \xda
+Y= \xdb
+%" \xdc
+JE \xdd
+JU \xde
+JA \xdf
+a= \xe0
+b= \xe1
+v= \xe2
+g= \xe3
+d= \xe4
+e= \xe5
+z% \xe6
+z= \xe7
+i= \xe8
+j= \xe9
+k= \xea
+l= \xeb
+m= \xec
+n= \xed
+o= \xee
+p= \xef
+r= \xf0
+s= \xf1
+t= \xf2
+u= \xf3
+f= \xf4
+h= \xf5
+c= \xf6
+c% \xf7
+s% \xf8
+sc \xf9
+=' \xfa
+y= \xfb
+%' \xfc
+je \xfd
+ju \xfe
+ja \xff
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/map.CP866 src/share/locale/colldef/map.CP866
--- src54-s/share/locale/colldef/map.CP866	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/map.CP866	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,175 @@
+# $FreeBSD$
+NU \x00
+SH \x01
+SX \x02
+EX \x03
+ET \x04
+EQ \x05
+AK \x06
+BL \x07
+BS \x08
+HT \x09
+LF \x0a
+VT \x0b
+FF \x0c
+CR \x0d
+SO \x0e
+SI \x0f
+DL \x10
+D1 \x11
+D2 \x12
+D3 \x13
+D4 \x14
+NK \x15
+SY \x16
+EB \x17
+CN \x18
+EM \x19
+SB \x1a
+EC \x1b
+FS \x1c
+GS \x1d
+RS \x1e
+US \x1f
+SP \x20
+Nb \x23
+DO \x24
+At \x40
+<( \x5b
+// \x5c
+)> \x5d
+'> \x5e
+'! \x60
+(! \x7b
+!! \x7c
+!) \x7d
+'? \x7e
+DT \x7f
+A= \x80
+B= \x81
+V= \x82
+G= \x83
+D= \x84
+E= \x85
+Z% \x86
+Z= \x87
+I= \x88
+J= \x89
+K= \x8a
+L= \x8b
+M= \x8c
+N= \x8d
+O= \x8e
+P= \x8f
+R= \x90
+S= \x91
+T= \x92
+U= \x93
+F= \x94
+H= \x95
+C= \x96
+C% \x97
+S% \x98
+Sc \x99
+=" \x9a
+Y= \x9b
+%" \x9c
+JE \x9d
+JU \x9e
+JA \x9f
+a= \xa0
+b= \xa1
+v= \xa2
+g= \xa3
+d= \xa4
+e= \xa5
+z% \xa6
+z= \xa7
+i= \xa8
+j= \xa9
+k= \xaa
+l= \xab
+m= \xac
+n= \xad
+o= \xae
+p= \xaf
+.S \xb0
+:S \xb1
+?S \xb2
+vv \xb3
+vl \xb4
+vL \xb5
+Vl \xb6
+Dl \xb7
+dL \xb8
+VL \xb9
+VV \xba
+LD \xbb
+UL \xbc
+Ul \xbd
+uL \xbe
+dl \xbf
+ur \xc0
+uh \xc1
+dh \xc2
+vr \xc3
+hh \xc4
+vh \xc5
+vR \xc6
+Vr \xc7
+UR \xc8
+DR \xc9
+UH \xca
+DH \xcb
+VR \xcc
+HH \xcd
+VH \xce
+uH \xcf
+Uh \xd0
+dH \xd1
+Dh \xd2
+Ur \xd3
+uR \xd4
+dR \xd5
+Dr \xd6
+Vh \xd7
+vH \xd8
+ul \xd9
+dr \xda
+FB \xdb
+LB \xdc
+lB \xdd
+RB \xde
+TB \xdf
+r= \xe0
+s= \xe1
+t= \xe2
+u= \xe3
+f= \xe4
+h= \xe5
+c= \xe6
+c% \xe7
+s% \xe8
+sc \xe9
+=' \xea
+y= \xeb
+%' \xec
+je \xed
+ju \xee
+ja \xef
+IO \xf0
+io \xf1
+IE \xf2
+ie \xf3
+YI \xf4
+yi \xf5
+V% \xf6
+v% \xf7
+DG \xf8
+Sb \xf9
+.M \xfa
+RT \xfb
+N0 \xfc
+Cu \xfd
+fS \xfe
+NS \xff
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/map.ISO8859-1 src/share/locale/colldef/map.ISO8859-1
--- src54-s/share/locale/colldef/map.ISO8859-1	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/map.ISO8859-1	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,174 @@
+NU \x00
+SH \x01
+SX \x02
+EX \x03
+ET \x04
+EQ \x05
+AK \x06
+BL \x07
+BS \x08
+HT \x09
+LF \x0a
+VT \x0b
+FF \x0c
+CR \x0d
+SO \x0e
+SI \x0f
+DL \x10
+D1 \x11
+D2 \x12
+D3 \x13
+D4 \x14
+NK \x15
+SY \x16
+EB \x17
+CN \x18
+EM \x19
+SB \x1a
+EC \x1b
+FS \x1c
+GS \x1d
+RS \x1e
+US \x1f
+SP \x20
+Nb \x23
+DO \x24
+At \x40
+<( \x5b
+// \x5c
+)> \x5d
+'> \x5e
+'! \x60
+(! \x7b
+!! \x7c
+!) \x7d
+'? \x7e
+DT \x7f
+PA \x80
+HO \x81
+BH \x82
+NH \x83
+IN \x84
+NL \x85
+SA \x86
+ES \x87
+HS \x88
+HJ \x89
+VS \x8a
+PD \x8b
+PU \x8c
+RI \x8d
+S2 \x8e
+S3 \x8f
+DC \x90
+P1 \x91
+P2 \x92
+TS \x93
+CC \x94
+MW \x95
+SG \x96
+EG \x97
+SS \x98
+GC \x99
+SC \x9a
+CI \x9b
+ST \x9c
+OC \x9d
+PM \x9e
+AC \x9f
+NS \xa0
+!I \xa1
+Ct \xa2
+Pd \xa3
+Cu \xa4
+Ye \xa5
+BB \xa6
+SE \xa7
+': \xa8
+Co \xa9
+-a \xaa
+<< \xab
+NO \xac
+-- \xad
+Rg \xae
+'m \xaf
+DG \xb0
++- \xb1
+2S \xb2
+3S \xb3
+'' \xb4
+My \xb5
+PI \xb6
+.M \xb7
+', \xb8
+1S \xb9
+-o \xba
+>> \xbb
+14 \xbc
+12 \xbd
+34 \xbe
+?I \xbf
+A! \xc0
+A' \xc1
+A> \xc2
+A? \xc3
+A: \xc4
+AA \xc5
+AE \xc6
+C, \xc7
+E! \xc8
+E' \xc9
+E> \xca
+E: \xcb
+I! \xcc
+I' \xcd
+I> \xce
+I: \xcf
+D- \xd0
+N? \xd1
+O! \xd2
+O' \xd3
+O> \xd4
+O? \xd5
+O: \xd6
+*X \xd7
+O/ \xd8
+U! \xd9
+U' \xda
+U> \xdb
+U: \xdc
+Y' \xdd
+TH \xde
+ss \xdf
+a! \xe0
+a' \xe1
+a> \xe2
+a? \xe3
+a: \xe4
+aa \xe5
+ae \xe6
+c, \xe7
+e! \xe8
+e' \xe9
+e> \xea
+e: \xeb
+i! \xec
+i' \xed
+i> \xee
+i: \xef
+d- \xf0
+n? \xf1
+o! \xf2
+o' \xf3
+o> \xf4
+o? \xf5
+o: \xf6
+-: \xf7
+o/ \xf8
+u! \xf9
+u' \xfa
+u> \xfb
+u: \xfc
+y' \xfd
+th \xfe
+y: \xff
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/map.ISO8859-13 src/share/locale/colldef/map.ISO8859-13
--- src54-s/share/locale/colldef/map.ISO8859-13	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/map.ISO8859-13	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,175 @@
+# $FreeBSD$
+NU \x00
+SH \x01
+SX \x02
+EX \x03
+ET \x04
+EQ \x05
+AK \x06
+BL \x07
+BS \x08
+HT \x09
+LF \x0a
+VT \x0b
+FF \x0c
+CR \x0d
+SO \x0e
+SI \x0f
+DL \x10
+D1 \x11
+D2 \x12
+D3 \x13
+D4 \x14
+NK \x15
+SY \x16
+EB \x17
+CN \x18
+EM \x19
+SB \x1a
+EC \x1b
+FS \x1c
+GS \x1d
+RS \x1e
+US \x1f
+SP \x20
+Nb \x23
+DO \x24
+At \x40
+<( \x5b
+// \x5c
+)> \x5d
+'> \x5e
+'! \x60
+(! \x7b
+!! \x7c
+!) \x7d
+'? \x7e
+DT \x7f
+PA \x80
+HO \x81
+BH \x82
+NH \x83
+IN \x84
+NL \x85
+SA \x86
+ES \x87
+HS \x88
+HJ \x89
+VS \x8a
+PD \x8b
+PU \x8c
+RI \x8d
+S2 \x8e
+S3 \x8f
+DC \x90
+P1 \x91
+P2 \x92
+TS \x93
+CC \x94
+MW \x95
+SG \x96
+EG \x97
+SS \x98
+GC \x99
+SC \x9a
+CI \x9b
+ST \x9c
+OC \x9d
+PM \x9e
+AC \x9f
+NS \xa0
+"" \xa1
+Ct \xa2
+Pd \xa3
+Cu \xa4
+,, \xa5
+BB \xa6
+SE \xa7
+O/ \xa8
+Co \xa9
+R, \xaa
+<< \xab
+NO \xac
+-- \xad
+Rg \xae
+AE \xaf
+DG \xb0
++- \xb1
+2S \xb2
+3S \xb3
+`` \xb4
+My \xb5
+PI \xb6
+.M \xb7
+o/ \xb8
+1S \xb9
+r, \xba
+>> \xbb
+14 \xbc
+12 \xbd
+34 \xbe
+ae \xbf
+A; \xc0
+I; \xc1
+A- \xc2
+C' \xc3
+A: \xc4
+AA \xc5
+E; \xc6
+E- \xc7
+C< \xc8
+E' \xc9
+Z' \xca
+E. \xcb
+G, \xcc
+K, \xcd
+I- \xce
+L, \xcf
+S< \xd0
+N' \xd1
+N, \xd2
+O' \xd3
+O- \xd4
+O? \xd5
+O: \xd6
+*X \xd7
+U; \xd8
+L/ \xd9
+S' \xda
+U- \xdb
+U: \xdc
+Z. \xdd
+Z< \xde
+ss \xdf
+a; \xe0
+i; \xe1
+a- \xe2
+c' \xe3
+a: \xe4
+aa \xe5
+e; \xe6
+e- \xe7
+c< \xe8
+e' \xe9
+z' \xea
+e. \xeb
+g, \xec
+k, \xed
+i- \xee
+l, \xef
+s< \xf0
+n' \xf1
+n, \xf2
+o' \xf3
+o- \xf4
+o? \xf5
+o: \xf6
+-: \xf7
+u; \xf8
+l/ \xf9
+s' \xfa
+u- \xfb
+u: \xfc
+z. \xfd
+z< \xfe
+'. \xff
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/map.ISO8859-15 src/share/locale/colldef/map.ISO8859-15
--- src54-s/share/locale/colldef/map.ISO8859-15	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/map.ISO8859-15	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,174 @@
+NU \x00
+SH \x01
+SX \x02
+EX \x03
+ET \x04
+EQ \x05
+AK \x06
+BL \x07
+BS \x08
+HT \x09
+LF \x0a
+VT \x0b
+FF \x0c
+CR \x0d
+SO \x0e
+SI \x0f
+DL \x10
+D1 \x11
+D2 \x12
+D3 \x13
+D4 \x14
+NK \x15
+SY \x16
+EB \x17
+CN \x18
+EM \x19
+SB \x1a
+EC \x1b
+FS \x1c
+GS \x1d
+RS \x1e
+US \x1f
+SP \x20
+Nb \x23
+DO \x24
+At \x40
+<( \x5b
+// \x5c
+)> \x5d
+'> \x5e
+'! \x60
+(! \x7b
+!! \x7c
+!) \x7d
+'? \x7e
+DT \x7f
+PA \x80
+HO \x81
+BH \x82
+NH \x83
+IN \x84
+NL \x85
+SA \x86
+ES \x87
+HS \x88
+HJ \x89
+VS \x8a
+PD \x8b
+PU \x8c
+RI \x8d
+S2 \x8e
+S3 \x8f
+DC \x90
+P1 \x91
+P2 \x92
+TS \x93
+CC \x94
+MW \x95
+SG \x96
+EG \x97
+SS \x98
+GC \x99
+SC \x9a
+CI \x9b
+ST \x9c
+OC \x9d
+PM \x9e
+AC \x9f
+NS \xa0
+!I \xa1
+Ct \xa2
+Pd \xa3
+Eu \xa4
+Ye \xa5
+S< \xa6
+SE \xa7
+s< \xa8
+Co \xa9
+-a \xaa
+<< \xab
+NO \xac
+-- \xad
+Rg \xae
+'m \xaf
+DG \xb0
++- \xb1
+2S \xb2
+3S \xb3
+Z< \xb4
+My \xb5
+PI \xb6
+.M \xb7
+z< \xb8
+1S \xb9
+-o \xba
+>> \xbb
+OE \xbc
+oe \xbd
+Y: \xbe
+?I \xbf
+A! \xc0
+A' \xc1
+A> \xc2
+A? \xc3
+A: \xc4
+AA \xc5
+AE \xc6
+C, \xc7
+E! \xc8
+E' \xc9
+E> \xca
+E: \xcb
+I! \xcc
+I' \xcd
+I> \xce
+I: \xcf
+D- \xd0
+N? \xd1
+O! \xd2
+O' \xd3
+O> \xd4
+O? \xd5
+O: \xd6
+*X \xd7
+O/ \xd8
+U! \xd9
+U' \xda
+U> \xdb
+U: \xdc
+Y' \xdd
+TH \xde
+ss \xdf
+a! \xe0
+a' \xe1
+a> \xe2
+a? \xe3
+a: \xe4
+aa \xe5
+ae \xe6
+c, \xe7
+e! \xe8
+e' \xe9
+e> \xea
+e: \xeb
+i! \xec
+i' \xed
+i> \xee
+i: \xef
+d- \xf0
+n? \xf1
+o! \xf2
+o' \xf3
+o> \xf4
+o? \xf5
+o: \xf6
+-: \xf7
+o/ \xf8
+u! \xf9
+u' \xfa
+u> \xfb
+u: \xfc
+y' \xfd
+th \xfe
+y: \xff
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/map.ISO8859-2 src/share/locale/colldef/map.ISO8859-2
--- src54-s/share/locale/colldef/map.ISO8859-2	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/map.ISO8859-2	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,174 @@
+NU \x00
+SH \x01
+SX \x02
+EX \x03
+ET \x04
+EQ \x05
+AK \x06
+BL \x07
+BS \x08
+HT \x09
+LF \x0a
+VT \x0b
+FF \x0c
+CR \x0d
+SO \x0e
+SI \x0f
+DL \x10
+D1 \x11
+D2 \x12
+D3 \x13
+D4 \x14
+NK \x15
+SY \x16
+EB \x17
+CN \x18
+EM \x19
+SB \x1a
+EC \x1b
+FS \x1c
+GS \x1d
+RS \x1e
+US \x1f
+SP \x20
+Nb \x23
+DO \x24
+At \x40
+<( \x5b
+// \x5c
+)> \x5d
+'> \x5e
+'! \x60
+(! \x7b
+!! \x7c
+!) \x7d
+'? \x7e
+DT \x7f
+PA \x80
+HO \x81
+BH \x82
+NH \x83
+IN \x84
+NL \x85
+SA \x86
+ES \x87
+HS \x88
+HJ \x89
+VS \x8a
+PD \x8b
+PU \x8c
+RI \x8d
+S2 \x8e
+S3 \x8f
+DC \x90
+P1 \x91
+P2 \x92
+TS \x93
+CC \x94
+MW \x95
+SG \x96
+EG \x97
+SS \x98
+GC \x99
+SC \x9a
+CI \x9b
+ST \x9c
+OC \x9d
+PM \x9e
+AC \x9f
+NS \xa0
+A; \xa1
+'( \xa2
+L/ \xa3
+Cu \xa4
+L< \xa5
+S' \xa6
+SE \xa7
+': \xa8
+S< \xa9
+S, \xaa
+T< \xab
+Z' \xac
+-- \xad
+Z< \xae
+Z. \xaf
+DG \xb0
+a; \xb1
+'; \xb2
+l/ \xb3
+'' \xb4
+l< \xb5
+s' \xb6
+'< \xb7
+', \xb8
+s< \xb9
+s, \xba
+t< \xbb
+z' \xbc
+'" \xbd
+z< \xbe
+z. \xbf
+R' \xc0
+A' \xc1
+A> \xc2
+A( \xc3
+A: \xc4
+L' \xc5
+C' \xc6
+C, \xc7
+C< \xc8
+E' \xc9
+E; \xca
+E: \xcb
+E< \xcc
+I' \xcd
+I> \xce
+D< \xcf
+D/ \xd0
+N' \xd1
+N< \xd2
+O' \xd3
+O> \xd4
+O" \xd5
+O: \xd6
+*X \xd7
+R< \xd8
+U0 \xd9
+U' \xda
+U" \xdb
+U: \xdc
+Y' \xdd
+T, \xde
+ss \xdf
+r' \xe0
+a' \xe1
+a> \xe2
+a( \xe3
+a: \xe4
+l' \xe5
+c' \xe6
+c, \xe7
+c< \xe8
+e' \xe9
+e; \xea
+e: \xeb
+e< \xec
+i' \xed
+i> \xee
+d< \xef
+d/ \xf0
+n' \xf1
+n< \xf2
+o' \xf3
+o> \xf4
+o" \xf5
+o: \xf6
+-: \xf7
+r< \xf8
+u0 \xf9
+u' \xfa
+u" \xfb
+u: \xfc
+y' \xfd
+t, \xfe
+'. \xff
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/map.ISO8859-4 src/share/locale/colldef/map.ISO8859-4
--- src54-s/share/locale/colldef/map.ISO8859-4	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/map.ISO8859-4	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,175 @@
+# $FreeBSD$
+NU \x00
+SH \x01
+SX \x02
+EX \x03
+ET \x04
+EQ \x05
+AK \x06
+BL \x07
+BS \x08
+HT \x09
+LF \x0a
+VT \x0b
+FF \x0c
+CR \x0d
+SO \x0e
+SI \x0f
+DL \x10
+D1 \x11
+D2 \x12
+D3 \x13
+D4 \x14
+NK \x15
+SY \x16
+EB \x17
+CN \x18
+EM \x19
+SB \x1a
+EC \x1b
+FS \x1c
+GS \x1d
+RS \x1e
+US \x1f
+SP \x20
+Nb \x23
+DO \x24
+At \x40
+<( \x5b
+// \x5c
+)> \x5d
+'> \x5e
+'! \x60
+(! \x7b
+!! \x7c
+!) \x7d
+'? \x7e
+DT \x7f
+PA \x80
+HO \x81
+BH \x82
+NH \x83
+IN \x84
+NL \x85
+SA \x86
+ES \x87
+HS \x88
+HJ \x89
+VS \x8a
+PD \x8b
+PU \x8c
+RI \x8d
+S2 \x8e
+S3 \x8f
+DC \x90
+P1 \x91
+P2 \x92
+TS \x93
+CC \x94
+MW \x95
+SG \x96
+EG \x97
+SS \x98
+GC \x99
+SC \x9a
+CI \x9b
+ST \x9c
+OC \x9d
+PM \x9e
+AC \x9f
+NS \xa0
+A; \xa1
+kk \xa2
+R, \xa3
+Cu \xa4
+I? \xa5
+L, \xa6
+SE \xa7
+': \xa8
+S< \xa9
+E- \xaa
+G, \xab
+T/ \xac
+-- \xad
+Z< \xae
+'m \xaf
+DG \xb0
+a; \xb1
+'; \xb2
+r, \xb3
+'' \xb4
+i? \xb5
+l, \xb6
+'< \xb7
+', \xb8
+s< \xb9
+e- \xba
+g, \xbb
+t/ \xbc
+NG \xbd
+z< \xbe
+ng \xbf
+A- \xc0
+A' \xc1
+A> \xc2
+A? \xc3
+A: \xc4
+AA \xc5
+AE \xc6
+I; \xc7
+C< \xc8
+E' \xc9
+E; \xca
+E: \xcb
+E. \xcc
+I' \xcd
+I> \xce
+I- \xcf
+D/ \xd0
+N, \xd1
+O- \xd2
+K, \xd3
+O> \xd4
+O? \xd5
+O: \xd6
+*X \xd7
+O/ \xd8
+U; \xd9
+U' \xda
+U> \xdb
+U: \xdc
+U? \xdd
+U- \xde
+ss \xdf
+a- \xe0
+a' \xe1
+a> \xe2
+a? \xe3
+a: \xe4
+aa \xe5
+ae \xe6
+i; \xe7
+c< \xe8
+e' \xe9
+e; \xea
+e: \xeb
+e. \xec
+i' \xed
+i> \xee
+i- \xef
+d/ \xf0
+n, \xf1
+o- \xf2
+k, \xf3
+o> \xf4
+o? \xf5
+o: \xf6
+-: \xf7
+o/ \xf8
+u; \xf9
+u' \xfa
+u> \xfb
+u: \xfc
+u? \xfd
+u- \xfe
+'. \xff
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/map.ISO8859-5 src/share/locale/colldef/map.ISO8859-5
--- src54-s/share/locale/colldef/map.ISO8859-5	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/map.ISO8859-5	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,175 @@
+# $FreeBSD$
+NU \x00
+SH \x01
+SX \x02
+EX \x03
+ET \x04
+EQ \x05
+AK \x06
+BL \x07
+BS \x08
+HT \x09
+LF \x0a
+VT \x0b
+FF \x0c
+CR \x0d
+SO \x0e
+SI \x0f
+DL \x10
+D1 \x11
+D2 \x12
+D3 \x13
+D4 \x14
+NK \x15
+SY \x16
+EB \x17
+CN \x18
+EM \x19
+SB \x1a
+EC \x1b
+FS \x1c
+GS \x1d
+RS \x1e
+US \x1f
+SP \x20
+Nb \x23
+DO \x24
+At \x40
+<( \x5b
+// \x5c
+)> \x5d
+'> \x5e
+'! \x60
+(! \x7b
+!! \x7c
+!) \x7d
+'? \x7e
+DT \x7f
+PA \x80
+HO \x81
+BH \x82
+NH \x83
+IN \x84
+NL \x85
+SA \x86
+ES \x87
+HS \x88
+HJ \x89
+VS \x8a
+PD \x8b
+PU \x8c
+RI \x8d
+S2 \x8e
+S3 \x8f
+DC \x90
+P1 \x91
+P2 \x92
+TS \x93
+CC \x94
+MW \x95
+SG \x96
+EG \x97
+SS \x98
+GC \x99
+SC \x9a
+CI \x9b
+ST \x9c
+OC \x9d
+PM \x9e
+AC \x9f
+NS \xa0
+IO \xa1
+D% \xa2
+G% \xa3
+IE \xa4
+DS \xa5
+II \xa6
+YI \xa7
+J% \xa8
+LJ \xa9
+NJ \xaa
+Ts \xab
+KJ \xac
+-- \xad
+V% \xae
+DZ \xaf
+A= \xb0
+B= \xb1
+V= \xb2
+G= \xb3
+D= \xb4
+E= \xb5
+Z% \xb6
+Z= \xb7
+I= \xb8
+J= \xb9
+K= \xba
+L= \xbb
+M= \xbc
+N= \xbd
+O= \xbe
+P= \xbf
+R= \xc0
+S= \xc1
+T= \xc2
+U= \xc3
+F= \xc4
+H= \xc5
+C= \xc6
+C% \xc7
+S% \xc8
+Sc \xc9
+=" \xca
+Y= \xcb
+%" \xcc
+JE \xcd
+JU \xce
+JA \xcf
+a= \xd0
+b= \xd1
+v= \xd2
+g= \xd3
+d= \xd4
+e= \xd5
+z% \xd6
+z= \xd7
+i= \xd8
+j= \xd9
+k= \xda
+l= \xdb
+m= \xdc
+n= \xdd
+o= \xde
+p= \xdf
+r= \xe0
+s= \xe1
+t= \xe2
+u= \xe3
+f= \xe4
+h= \xe5
+c= \xe6
+c% \xe7
+s% \xe8
+sc \xe9
+=' \xea
+y= \xeb
+%' \xec
+je \xed
+ju \xee
+ja \xef
+N0 \xf0
+io \xf1
+d% \xf2
+g% \xf3
+ie \xf4
+ds \xf5
+ii \xf6
+yi \xf7
+j% \xf8
+lj \xf9
+nj \xfa
+ts \xfb
+kj \xfc
+SE \xfd
+v% \xfe
+dz \xff
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/map.ISO8859-7 src/share/locale/colldef/map.ISO8859-7
--- src54-s/share/locale/colldef/map.ISO8859-7	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/map.ISO8859-7	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,175 @@
+# $FreeBSD$
+NU \x00
+SH \x01
+SX \x02
+EX \x03
+ET \x04
+EQ \x05
+AK \x06
+BL \x07
+BS \x08
+HT \x09
+LF \x0a
+VT \x0b
+FF \x0c
+CR \x0d
+SO \x0e
+SI \x0f
+DL \x10
+D1 \x11
+D2 \x12
+D3 \x13
+D4 \x14
+NK \x15
+SY \x16
+EB \x17
+CN \x18
+EM \x19
+SB \x1a
+EC \x1b
+FS \x1c
+GS \x1d
+RS \x1e
+US \x1f
+SP \x20
+Nb \x23
+DO \x24
+At \x40
+<( \x5b
+// \x5c
+)> \x5d
+'> \x5e
+'! \x60
+(! \x7b
+!! \x7c
+!) \x7d
+'? \x7e
+DT \x7f
+PA \x80
+HO \x81
+BH \x82
+NH \x83
+IN \x84
+NL \x85
+SA \x86
+ES \x87
+HS \x88
+HJ \x89
+VS \x8a
+PD \x8b
+PU \x8c
+RI \x8d
+S2 \x8e
+S3 \x8f
+DC \x90
+P1 \x91
+P2 \x92
+TS \x93
+CC \x94
+MW \x95
+SG \x96
+EG \x97
+SS \x98
+GC \x99
+SC \x9a
+CI \x9b
+ST \x9c
+OC \x9d
+PM \x9e
+AC \x9f
+NS \xa0
+9' \xa1
+'9 \xa2
+Pd \xa3
+IE \xa4
+DS \xa5
+BB \xa6
+SE \xa7
+': \xa8
+Co \xa9
+NJ \xaa
+<< \xab
+NO \xac
+-- \xad
+V% \xae
+-M \xaf
+DG \xb0
++- \xb1
+2S \xb2
+3S \xb3
+'' \xb4
+'% \xb5
+A% \xb6
+.M \xb7
+E% \xb8
+Y% \xb9
+I% \xba
+>> \xbb
+O% \xbc
+12 \xbd
+U% \xbe
+W% \xbf
+i3 \xc0
+A* \xc1
+B* \xc2
+G* \xc3
+D* \xc4
+E* \xc5
+Z* \xc6
+Y* \xc7
+H* \xc8
+I* \xc9
+K* \xca
+L* \xcb
+M* \xcc
+N* \xcd
+C* \xce
+O* \xcf
+P* \xd0
+R* \xd1
+v= \xd2
+S* \xd3
+T* \xd4
+U* \xd5
+F* \xd6
+X* \xd7
+Q* \xd8
+W* \xd9
+J* \xda
+V* \xdb
+a% \xdc
+e% \xdd
+y% \xde
+i% \xdf
+u3 \xe0
+a* \xe1
+b* \xe2
+g* \xe3
+d* \xe4
+e* \xe5
+z* \xe6
+y* \xe7
+h* \xe8
+i* \xe9
+k* \xea
+l* \xeb
+m* \xec
+n* \xed
+c* \xee
+o* \xef
+p* \xf0
+r* \xf1
+*s \xf2
+s* \xf3
+t* \xf4
+u* \xf5
+f* \xf6
+x* \xf7
+q* \xf8
+w* \xf9
+j* \xfa
+v* \xfb
+o% \xfc
+u% \xfd
+w% \xfe
+dz \xff
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/map.ISO8859-9 src/share/locale/colldef/map.ISO8859-9
--- src54-s/share/locale/colldef/map.ISO8859-9	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/map.ISO8859-9	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,175 @@
+# $FreeBSD$
+NU \x00
+SH \x01
+SX \x02
+EX \x03
+ET \x04
+EQ \x05
+AK \x06
+BL \x07
+BS \x08
+HT \x09
+LF \x0a
+VT \x0b
+FF \x0c
+CR \x0d
+SO \x0e
+SI \x0f
+DL \x10
+D1 \x11
+D2 \x12
+D3 \x13
+D4 \x14
+NK \x15
+SY \x16
+EB \x17
+CN \x18
+EM \x19
+SB \x1a
+EC \x1b
+FS \x1c
+GS \x1d
+RS \x1e
+US \x1f
+SP \x20
+Nb \x23
+DO \x24
+At \x40
+<( \x5b
+// \x5c
+)> \x5d
+'> \x5e
+'! \x60
+(! \x7b
+!! \x7c
+!) \x7d
+'? \x7e
+DT \x7f
+PA \x80
+HO \x81
+BH \x82
+NH \x83
+IN \x84
+NL \x85
+SA \x86
+ES \x87
+HS \x88
+HJ \x89
+VS \x8a
+PD \x8b
+PU \x8c
+RI \x8d
+S2 \x8e
+S3 \x8f
+DC \x90
+P1 \x91
+P2 \x92
+TS \x93
+CC \x94
+MW \x95
+SG \x96
+EG \x97
+SS \x98
+GC \x99
+SC \x9a
+CI \x9b
+ST \x9c
+OC \x9d
+PM \x9e
+AC \x9f
+NS \xa0
+!I \xa1
+Ct \xa2
+Pd \xa3
+Cu \xa4
+Ye \xa5
+BB \xa6
+SE \xa7
+': \xa8
+Co \xa9
+-a \xaa
+<< \xab
+NO \xac
+-- \xad
+Rg \xae
+'m \xaf
+DG \xb0
++- \xb1
+2S \xb2
+3S \xb3
+'' \xb4
+My \xb5
+PI \xb6
+.M \xb7
+', \xb8
+1S \xb9
+-o \xba
+>> \xbb
+14 \xbc
+12 \xbd
+34 \xbe
+?I \xbf
+A! \xc0
+A' \xc1
+A> \xc2
+A? \xc3
+A: \xc4
+AA \xc5
+AE \xc6
+C, \xc7
+E! \xc8
+E' \xc9
+E> \xca
+E: \xcb
+I! \xcc
+I' \xcd
+I> \xce
+I: \xcf
+G( \xd0
+N? \xd1
+O! \xd2
+O' \xd3
+O> \xd4
+O? \xd5
+O: \xd6
+*X \xd7
+O/ \xd8
+U! \xd9
+U' \xda
+U> \xdb
+U: \xdc
+I. \xdd
+S, \xde
+ss \xdf
+a! \xe0
+a' \xe1
+a> \xe2
+a? \xe3
+a: \xe4
+aa \xe5
+ae \xe6
+c, \xe7
+e! \xe8
+e' \xe9
+e; \xea
+e: \xeb
+e. \xec
+i' \xed
+i> \xee
+i- \xef
+g( \xf0
+n? \xf1
+o! \xf2
+o' \xf3
+o> \xf4
+o? \xf5
+o: \xf6
+-: \xf7
+o/ \xf8
+u! \xf9
+u' \xfa
+u> \xfb
+u: \xfc
+i. \xfd
+s, \xfe
+y: \xff
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/map.KOI8-R src/share/locale/colldef/map.KOI8-R
--- src54-s/share/locale/colldef/map.KOI8-R	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/map.KOI8-R	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,175 @@
+# $FreeBSD$
+NU \x00
+SH \x01
+SX \x02
+EX \x03
+ET \x04
+EQ \x05
+AK \x06
+BL \x07
+BS \x08
+HT \x09
+LF \x0a
+VT \x0b
+FF \x0c
+CR \x0d
+SO \x0e
+SI \x0f
+DL \x10
+D1 \x11
+D2 \x12
+D3 \x13
+D4 \x14
+NK \x15
+SY \x16
+EB \x17
+CN \x18
+EM \x19
+SB \x1a
+EC \x1b
+FS \x1c
+GS \x1d
+RS \x1e
+US \x1f
+SP \x20
+Nb \x23
+DO \x24
+At \x40
+<( \x5b
+// \x5c
+)> \x5d
+'> \x5e
+'! \x60
+(! \x7b
+!! \x7c
+!) \x7d
+'? \x7e
+DT \x7f
+hh \x80
+vv \x81
+dr \x82
+dl \x83
+ur \x84
+ul \x85
+vr \x86
+vl \x87
+dh \x88
+uh \x89
+vh \x8a
+TB \x8b
+LB \x8c
+FB \x8d
+lB \x8e
+RB \x8f
+.S \x90
+:S \x91
+?S \x92
+Iu \x93
+fS \x94
+Sb \x95
+RT \x96
+?2 \x97
+=< \x98
+>= \x99
+NS \x9a
+Il \x9b
+DG \x9c
+2S \x9d
+.M \x9e
+-: \x9f
+HH \xa0
+VV \xa1
+dR \xa2
+io \xa3
+Dr \xa4
+DR \xa5
+dL \xa6
+Dl \xa7
+LD \xa8
+uR \xa9
+Ur \xaa
+UR \xab
+uL \xac
+Ul \xad
+UL \xae
+vR \xaf
+Vr \xb0
+VR \xb1
+vL \xb2
+IO \xb3
+Vl \xb4
+VL \xb5
+dH \xb6
+Dh \xb7
+DH \xb8
+uH \xb9
+Uh \xba
+UH \xbb
+vH \xbc
+Vh \xbd
+VH \xbe
+Co \xbf
+ju \xc0
+a= \xc1
+b= \xc2
+c= \xc3
+d= \xc4
+e= \xc5
+f= \xc6
+g= \xc7
+h= \xc8
+i= \xc9
+j= \xca
+k= \xcb
+l= \xcc
+m= \xcd
+n= \xce
+o= \xcf
+p= \xd0
+ja \xd1
+r= \xd2
+s= \xd3
+t= \xd4
+u= \xd5
+z% \xd6
+v= \xd7
+%' \xd8
+y= \xd9
+z= \xda
+s% \xdb
+je \xdc
+sc \xdd
+c% \xde
+=' \xdf
+JU \xe0
+A= \xe1
+B= \xe2
+C= \xe3
+D= \xe4
+E= \xe5
+F= \xe6
+G= \xe7
+H= \xe8
+I= \xe9
+J= \xea
+K= \xeb
+L= \xec
+M= \xed
+N= \xee
+O= \xef
+P= \xf0
+JA \xf1
+R= \xf2
+S= \xf3
+T= \xf4
+U= \xf5
+Z% \xf6
+V= \xf7
+%" \xf8
+Y= \xf9
+Z= \xfa
+S% \xfb
+JE \xfc
+Sc \xfd
+C% \xfe
+=" \xff
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/map.KOI8-U src/share/locale/colldef/map.KOI8-U
--- src54-s/share/locale/colldef/map.KOI8-U	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/map.KOI8-U	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,175 @@
+# $FreeBSD$
+NU \x00
+SH \x01
+SX \x02
+EX \x03
+ET \x04
+EQ \x05
+AK \x06
+BL \x07
+BS \x08
+HT \x09
+LF \x0a
+VT \x0b
+FF \x0c
+CR \x0d
+SO \x0e
+SI \x0f
+DL \x10
+D1 \x11
+D2 \x12
+D3 \x13
+D4 \x14
+NK \x15
+SY \x16
+EB \x17
+CN \x18
+EM \x19
+SB \x1a
+EC \x1b
+FS \x1c
+GS \x1d
+RS \x1e
+US \x1f
+SP \x20
+Nb \x23
+DO \x24
+At \x40
+<( \x5b
+// \x5c
+)> \x5d
+'> \x5e
+'! \x60
+(! \x7b
+!! \x7c
+!) \x7d
+'? \x7e
+DT \x7f
+hh \x80
+vv \x81
+dr \x82
+dl \x83
+ur \x84
+ul \x85
+vr \x86
+vl \x87
+dh \x88
+uh \x89
+vh \x8a
+TB \x8b
+LB \x8c
+FB \x8d
+lB \x8e
+RB \x8f
+.S \x90
+:S \x91
+?S \x92
+Iu \x93
+fS \x94
+Sb \x95
+RT \x96
+?2 \x97
+=< \x98
+>= \x99
+NS \x9a
+Il \x9b
+DG \x9c
+2S \x9d
+.M \x9e
+-: \x9f
+HH \xa0
+VV \xa1
+dR \xa2
+io \xa3
+ie \xa4
+DR \xa5
+ii \xa6
+yi \xa7
+LD \xa8
+uR \xa9
+Ur \xaa
+UR \xab
+uL \xac
+g3 \xad
+UL \xae
+vR \xaf
+Vr \xb0
+VR \xb1
+vL \xb2
+IO \xb3
+IE \xb4
+VL \xb5
+II \xb6
+YI \xb7
+DH \xb8
+uH \xb9
+Uh \xba
+UH \xbb
+vH \xbc
+G3 \xbd
+VH \xbe
+Co \xbf
+ju \xc0
+a= \xc1
+b= \xc2
+c= \xc3
+d= \xc4
+e= \xc5
+f= \xc6
+g= \xc7
+h= \xc8
+i= \xc9
+j= \xca
+k= \xcb
+l= \xcc
+m= \xcd
+n= \xce
+o= \xcf
+p= \xd0
+ja \xd1
+r= \xd2
+s= \xd3
+t= \xd4
+u= \xd5
+z% \xd6
+v= \xd7
+%' \xd8
+y= \xd9
+z= \xda
+s% \xdb
+je \xdc
+sc \xdd
+c% \xde
+=' \xdf
+JU \xe0
+A= \xe1
+B= \xe2
+C= \xe3
+D= \xe4
+E= \xe5
+F= \xe6
+G= \xe7
+H= \xe8
+I= \xe9
+J= \xea
+K= \xeb
+L= \xec
+M= \xed
+N= \xee
+O= \xef
+P= \xf0
+JA \xf1
+R= \xf2
+S= \xf3
+T= \xf4
+U= \xf5
+Z% \xf6
+V= \xf7
+%" \xf8
+Y= \xf9
+Z= \xfa
+S% \xfb
+JE \xfc
+Sc \xfd
+C% \xfe
+=" \xff
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/map.PT154 src/share/locale/colldef/map.PT154
--- src54-s/share/locale/colldef/map.PT154	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/map.PT154	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,258 @@
+# $FreeBSD$
+# Birsh T <birsh@mail.kz>, 2003-12-12
+NU \x00
+SH \x01
+SX \x02
+EX \x03
+ET \x04
+EQ \x05
+AK \x06
+BL \x07
+BS \x08
+HT \x09
+LF \x0a
+VT \x0b
+FF \x0c
+CR \x0d
+SO \x0e
+SI \x0f
+DL \x10
+D1 \x11
+D2 \x12
+D3 \x13
+D4 \x14
+NK \x15
+SY \x16
+EB \x17
+CN \x18
+EM \x19
+SB \x1a
+EC \x1b
+FS \x1c
+GS \x1d
+RS \x1e
+US \x1f
+SP \x20
+! \x21
+" \x22
+Nb \x23
+DO \x24
+% \x25
+& \x26
+' \x27
+( \x28
+) \x29
+* \x2a
++ \x2b
+, \x2c
+- \x2d
+. \x2e
+/ \x2f
+0 \x30
+1 \x31
+2 \x32
+3 \x33
+4 \x34
+5 \x35
+6 \x36
+7 \x37
+8 \x38
+9 \x39
+: \x3a
+; \x3b
+< \x3c
+= \x3d
+> \x3e
+? \x3f
+At \x40
+A \x41
+B \x42
+C \x43
+D \x44
+E \x45
+F \x46
+G \x47
+H \x48
+I \x49
+J \x4a
+K \x4b
+L \x4c
+M \x4d
+N \x4e
+O \x4f
+P \x50
+Q \x51
+R \x52
+S \x53
+T \x54
+U \x55
+V \x56
+W \x57
+X \x58
+Y \x59
+Z \x5a
+<( \x5b
+// \x5c
+)> \x5d
+'> \x5e
+_ \x5f
+'! \x60
+a \x61
+b \x62
+c \x63
+d \x64
+e \x65
+f \x66
+g \x67
+h \x68
+i \x69
+j \x6a
+k \x6b
+l \x6c
+m \x6d
+n \x6e
+o \x6f
+p \x70
+q \x71
+r \x72
+s \x73
+t \x74
+u \x75
+v \x76
+w \x77
+x \x78
+y \x79
+z \x7a
+(! \x7b
+!! \x7c
+!) \x7d
+'? \x7e
+DT \x7f
+Z. \x80
+G- \x81
+U' \x82
+g- \x83
+:9 \x84
+.3 \x85
+C. \x86
+Y" \x87
+H% \x88
+y" \x89
+BK \x8a
+I' \x8b
+N% \x8c
+K% \x8d
+HH \x8e
+C! \x8f
+z. \x90
+'6 \x91
+'9 \x92
+"6 \x93
+"9 \x94
+sb \x95
+-N \x96
+-M \x97
+h% \x98
+c. \x99
+bk \x9a
+i' \x9b
+n% \x9c
+k% \x9d
+hh \x9e
+c! \x9f
+NS \xa0
+V% \xa1
+v% \xa2
+J% \xa3
+O- \xa4
+X% \xa5
+Y- \xa6
+SE \xa7
+IO \xa8
+Co \xa9
+HW \xaa
+<< \xab
+NO \xac
+u' \xad
+Rg \xae
+K! \xaf
+DG \xb0
+y- \xb1
+II \xb2
+ii \xb3
+x% \xb4
+o- \xb5
+PI \xb6
+.M \xb7
+io \xb8
+N0 \xb9
+hw \xba
+>> \xbb
+j% \xbc
+DS \xbd
+ds \xbe
+k! \xbf
+A= \xc0
+B= \xc1
+V= \xc2
+G= \xc3
+D= \xc4
+E= \xc5
+Z% \xc6
+Z= \xc7
+I= \xc8
+J= \xc9
+K= \xca
+L= \xcb
+M= \xcc
+N= \xcd
+O= \xce
+P= \xcf
+R= \xd0
+S= \xd1
+T= \xd2
+U= \xd3
+F= \xd4
+H= \xd5
+C= \xd6
+C% \xd7
+S% \xd8
+Sc \xd9
+=" \xda
+Y= \xdb
+%" \xdc
+JE \xdd
+JU \xde
+JA \xdf
+a= \xe0
+b= \xe1
+v= \xe2
+g= \xe3
+d= \xe4
+e= \xe5
+z% \xe6
+z= \xe7
+i= \xe8
+j= \xe9
+k= \xea
+l= \xeb
+m= \xec
+n= \xed
+o= \xee
+p= \xef
+r= \xf0
+s= \xf1
+t= \xf2
+u= \xf3
+f= \xf4
+h= \xf5
+c= \xf6
+c% \xf7
+s% \xf8
+sc \xf9
+=' \xfa
+y= \xfb
+%' \xfc
+je \xfd
+ju \xfe
+ja \xff
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/no.ISO8859-1.src src/share/locale/colldef/no.ISO8859-1.src
--- src54-s/share/locale/colldef/no.ISO8859-1.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/no.ISO8859-1.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,42 @@
+# Norwegian/ISO 8859-1 (backward compatible with ASCII)
+#
+# (Based on the Swedish sv_SE.ISO_8859-1)
+#
+# $FreeBSD$
+#
+charmap map.ISO8859-1
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+#
+	<NS>;<SP>;!;<!I>;\";<<<>;</>/>>;<Nb>;\
+	<Cu>;<Ct>;<DO>;<Pd>;<Ye>;\
+	%;&;<',>;';\(;\);*;+;<+->;<-:>;<*X>;\,;<-->;-;.;/;\
+# digits
+	(0,<14>,<12>,<34>);(1,<1S>);(2,<2S>);(3,<3S>);4;...;9;\
+#
+	:;\;;\<;=;>;?;<?I>;<SE>;<PI>;<Co>;<Rg>;<At>;\
+# capital
+	(A,<A'>,<A!>,<A/>>,<A?>);\
+	B;(C,<C,>);(D,<D->);(E,<E'>,<E!>,<E/>>,<E:>);\
+	F;G;H;(I,<I'>,<I!>,<I/>>,<I:>);\
+	J;...;M;(N,<N?>);(O,<O'>,<O!>,<O?>,<O/>>);\
+	P;...;T;(U,<U'>,<U!>,<U/>>);\
+	V;W;X;(Y,<Y'>,<U:>);Z;\
+	<TH>;\
+	(<AE>,<A:>);(<O//>,<O:>);<AA>;\
+#
+	[;\\;];^;<':>;_;<'m>;<''>;`;\
+# small
+	(a,<a'>,<a!>,<a/>>,<a?>);\
+	b;(c,<c,>);(d,<d->);(e,<e'>,<e!>,<e/>>,<e:>);\
+	f;g;h;(i,<i'>,<i!>,<i/>>,<i:>);\
+	j;...;m;(n,<n?>);(o,<o'>,<o!>,<o?>,<o/>>);\
+	p;...;t;(u,<u'>,<u!>,<u/>>);\
+	v;w;x;(y,<y'>,<y:>,<u:>);z;\
+	<th>;\
+	(<ae>,<a:>);(<o//>,<o:>);<aa>;\
+#
+	\{;<NO>;|;<BB>;\};~;<.M>;<DG>;<My>;<DT>;\
+# remains
+	<-a>;<ss>;<-o>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/no.ISO8859-15.src src/share/locale/colldef/no.ISO8859-15.src
--- src54-s/share/locale/colldef/no.ISO8859-15.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/no.ISO8859-15.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,42 @@
+# Norwegian/ISO 8859-15 (backward compatible with ASCII)
+#
+# (Based on the Swedish sv_SE.ISO_8859-15)
+#
+# $FreeBSD$
+#
+charmap map.ISO8859-15
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+#
+	<NS>;<SP>;!;<!I>;\";<<<>;</>/>>;<Nb>;\
+	<Eu>;<Ct>;<DO>;<Pd>;<Ye>;\
+	%;&;';\(;\);*;+;<+->;<-:>;<*X>;\,;<-->;-;.;/;\
+# digits
+	0;(1,<1S>);(2,<2S>);(3,<3S>);4;...;9;\
+#
+	:;\;;\<;=;>;?;<?I>;<SE>;<PI>;<Co>;<Rg>;<At>;\
+# capital
+	(A,<A'>,<A!>,<A/>>,<A?>);\
+	B;(C,<C,>);(D,<D->);(E,<E'>,<E!>,<E/>>,<E:>);\
+	F;G;H;(I,<I'>,<I!>,<I/>>,<I:>);\
+	J;...;M;(N,<N?>);(O,<O'>,<O!>,<O?>,<O/>>,<OE>);\
+	P;Q;R;(S,<S<>);T;(U,<U'>,<U!>,<U/>>);\
+	V;W;X;(Y,<Y'>,<Y:>,<U:>);(Z,<Z<>);\
+	<TH>;\
+	(<AE>,<A:>);(<O//>,<O:>);<AA>;\
+#
+	[;\\;];^;_;<'m>;`;\
+# small
+	(a,<a'>,<a!>,<a/>>,<a?>);\
+	b;(c,<c,>);(d,<d->);(e,<e'>,<e!>,<e/>>,<e:>);\
+	f;g;h;(i,<i'>,<i!>,<i/>>,<i:>);\
+	j;...;m;(n,<n?>);(o,<o'>,<o!>,<o?>,<o/>>,<oe>);\
+	p;q;r;(s,<s<>,<ss>);t;(u,<u'>,<u!>,<u/>>);\
+	v;w;x;(y,<y'>,<y:>,<u:>);(z,<z<>);\
+	<th>;\
+	(<ae>,<a:>);(<o//>,<o:>);<aa>;\
+#
+	\{;<NO>;|;\};~;<.M>;<DG>;<My>;<DT>;\
+# remains
+	<-a>;<-o>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/pl.ISO8859-2.src src/share/locale/colldef/pl.ISO8859-2.src
--- src54-s/share/locale/colldef/pl.ISO8859-2.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/pl.ISO8859-2.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,37 @@
+# Latin2 (for Polish) by Michal Pasternak <michal@pasternak.w.lub.pl> 
+# with much help from Adam Liberacki and Jan Srzednicki
+#
+# $FreeBSD$
+#
+charmap map.ISO8859-2
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+# 
+	<NS>;<SP>;!;\";<Nb>;\
+	<Cu>;<DO>;\
+	%;&;<',>;<';>;';\(;\);*;+;<-:>;<*X>;\,;<-->;-;.;<'.>;/;\
+# digits
+	0;...;9;\
+# 
+	:;\;;\<;=;>;?;<SE>;<At>;\
+# capital
+	(A,<A'>,<A/>>,<A:>,<A(>);\
+	<A;>;B;(C,<C,>,<C<>);<C'>;(D,<D<>,<D//>);(E,<E'>,<E:>,<E<>);<E;>;\
+	F;G;H;(I,<I'>,<I/>>);\
+	J;K;(L,<L<>,<L'>);<L//>;M;(N,<N<>);<N'>;(O,<O/>>,<O:>,<O">);<O'>;\
+	P;Q;(R,<R'>,<R<>);(S,<S<>,<S,>);<S'>;(T,<T<>,<T,>);\
+	(U,<U'>,<U:>,<U0>,<U">);\
+	V;W;X;(Y,<Y'>);(Z,<Z<>);<Z'>;<Z.>;\
+#
+	[;\\;];^;<':>;<'">;<'<>;<'(>;_;<''>;`;\
+# small
+	(a,<a'>,<a/>>,<a:>,<a(>);<a;>;\
+	b;(c,<c,>,<c<>);<c'>;(d,<d<>,<d//>);(e,<e'>,<e:>,<e<>);<e;>;\
+	f;g;h;(i,<i'>,<i/>>);\
+	j;k;(l,<l<>,<l'>);<l//>;m;(n,<n<>);<n'>;(o,<o/>>,<o:>,<o">);<o'>;\
+	p;q;(r,<r'>,<r<>);(s,<s<>,<s,>,<ss>);<s'>;(t,<t<>,<t,>);\
+	(u,<u'>,<u:>,<u0>,<u">);\
+	v;w;x;(y,<y'>);(z,<z<>);<z'>;<z.>;\
+#
+	\{;|;\};~;<DG>;<DT>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/sl.ISO8859-2.src src/share/locale/colldef/sl.ISO8859-2.src
--- src54-s/share/locale/colldef/sl.ISO8859-2.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/sl.ISO8859-2.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,39 @@
+#
+# Slovenian collation sequence
+# 2003-03-20 simon.zekar@literal.si
+#
+# $FreeBSD$
+#
+
+charmap map.ISO8859-2
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+# 
+	<NS>;<SP>;!;\";<Nb>;\
+	<Cu>;<DO>;\
+	%;&;<',>;<';>;';\(;\);*;+;<-:>;<*X>;\,;<-->;-;.;<'.>;/;\
+# digits
+	0;...;9;\
+# 
+	:;\;;\<;=;>;?;<SE>;<At>;\
+# capital
+	(A,<A'>,<A/>>,<A:>,<A;>,<A(>);\
+	B;(C,<C,>);<C<>;<C'>;(D,<D<>);<D//>;(E,<E'>,<E:>,<E;>,<E<>);\
+	F;G;H;(I,<I'>,<I/>>);\
+	J;K;(L,<L//>,<L<>,<L'>);M;(N,<N'>,<N<>);(O,<O'>,<O/>>,<O:>,<O">);\
+	P;Q;(R,<R'>,<R<>);(S,<S'>,<S,>);<S<>;(T,<T<>,<T,>);\
+	(U,<U'>,<U:>,<U0>,<U">);\
+	V;W;X;(Y,<Y'>);(Z,<Z'>,<Z.>);<Z<>;\
+#
+	[;\\;];^;<':>;<'">;<'<>;<'(>;_;<''>;`;\
+# small
+	(a,<a'>,<a/>>,<a:>,<a;>,<a(>);\
+	b;(c,<c,>);<c<>;<c'>;(d,<d<>);<d//>;(e,<e'>,<e:>,<e;>,<e<>);\
+	f;g;h;(i,<i'>,<i/>>);\
+	j;k;(l,<l//>,<l<>,<l'>);m;(n,<n'>,<n<>);(o,<o'>,<o/>>,<o:>,<o">);\
+	p;q;(r,<r'>,<r<>);(s,<s'>,<s,>,<ss>);<s<>;(t,<t<>,<t,>);\
+	(u,<u'>,<u:>,<u0>,<u">);\
+	v;w;x;(y,<y'>);(z,<z'>,<z.>);<z<>;\
+#
+	\{;|;\};~;<DG>;<DT>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/sv.ISO8859-1.src src/share/locale/colldef/sv.ISO8859-1.src
--- src54-s/share/locale/colldef/sv.ISO8859-1.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/sv.ISO8859-1.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,40 @@
+# Swedish/ISO 8859-1 (backward compatible with ASCII)
+#
+# (Based on the German de_DE.ISO_8859-1)
+#
+# $FreeBSD$
+#
+charmap map.ISO8859-1
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+#
+	<NS>;<SP>;!;<!I>;\";<<<>;</>/>>;<Nb>;\
+	<Cu>;<Ct>;<DO>;<Pd>;<Ye>;\
+	%;&;<',>;';\(;\);*;+;<+->;<-:>;<*X>;\,;<-->;-;.;/;\
+# digits
+	(0,<14>,<12>,<34>);(1,<1S>);(2,<2S>);(3,<3S>);4;...;9;\
+#
+	:;\;;\<;=;>;?;<?I>;<SE>;<PI>;<Co>;<Rg>;<At>;\
+# capital
+	(A,<A'>,<A!>,<A/>>,<A?>);\
+	B;(C,<C,>);(D,<D->);(E,<E'>,<E!>,<E/>>,<E:>);\
+	F;G;H;(I,<I'>,<I!>,<I/>>,<I:>);\
+	J;...;M;(N,<N?>);(O,<O'>,<O!>,<O?>,<O/>>);\
+	P;...;T;(U,<U'>,<U!>,<U/>>,<U:>);\
+	V;W;X;(Y,<Y'>);Z;<AA>;(<A:>,<AE>);(<O:>,<O//>);\
+	<TH>;\
+#
+	[;\\;];^;<':>;_;<'m>;<''>;`;\
+# small
+	(a,<a'>,<a!>,<a/>>,<a?>);\
+	b;(c,<c,>);(d,<d->);(e,<e'>,<e!>,<e/>>,<e:>);\
+	f;g;h;(i,<i'>,<i!>,<i/>>,<i:>);\
+	j;...;m;(n,<n?>);(o,<o'>,<o!>,<o?>,<o/>>);\
+	p;...;t;(u,<u'>,<u!>,<u/>>,<u:>);\
+	v;w;x;(y,<y'>,<y:>);z;<aa>;(<a:>,<ae>);(<o:>,<o//>);\
+	<th>;\
+#
+	\{;<NO>;|;<BB>;\};~;<.M>;<DG>;<My>;<DT>;\
+# remains
+	<-a>;<ss>;<-o>
diff -ruN -x obj -x CVS -x *~ -x *patch -x *orig src54-s/share/locale/colldef/sv.ISO8859-15.src src/share/locale/colldef/sv.ISO8859-15.src
--- src54-s/share/locale/colldef/sv.ISO8859-15.src	Wed Dec 31 19:00:00 1969
+++ src/share/locale/colldef/sv.ISO8859-15.src	Sat Oct 12 13:23:41 2013
@@ -0,0 +1,40 @@
+# Swedish/DIS 8859-15 (backward compatible with ASCII)
+#
+# (Based on the German de_DE.ISO_8859-15)
+#
+# $FreeBSD$
+#
+charmap map.ISO8859-15
+order \
+# controls
+	<NU>;...;<US>;<PA>;...;<AC>;\
+#
+	<NS>;<SP>;!;<!I>;\";<<<>;</>/>>;<Nb>;\
+	<Eu>;<Ct>;<DO>;<Pd>;<Ye>;\
+	%;&;';\(;\);*;+;<+->;<-:>;<*X>;\,;<-->;-;.;/;\
+# digits
+	0;(1,<1S>);(2,<2S>);(3,<3S>);4;...;9;\
+#
+	:;\;;\<;=;>;?;<?I>;<SE>;<PI>;<Co>;<Rg>;<At>;\
+# capital
+	(A,<A'>,<A!>,<A/>>,<A?>);\
+	B;(C,<C,>);(D,<D->);(E,<E'>,<E!>,<E/>>,<E:>);\
+	F;G;H;(I,<I'>,<I!>,<I/>>,<I:>);\
+	J;...;M;(N,<N?>);(O,<O'>,<O!>,<O?>,<O/>>,<OE>);\
+	P;Q;R;(S,<S<>);T;(U,<U'>,<U!>,<U/>>,<U:>);\
+	V;W;X;(Y,<Y'>,<Y:>);(Z,<Z<>);<AA>;(<A:>,<AE>);(<O:>,<O//>);\
+	<TH>;\
+#
+	[;\\;];^;_;<'m>;`;\
+# small
+	(a,<a'>,<a!>,<a/>>,<a?>);\
+	b;(c,<c,>);(d,<d->);(e,<e'>,<e!>,<e/>>,<e:>);\
+	f;g;h;(i,<i'>,<i!>,<i/>>,<i:>);\
+	j;...;m;(n,<n?>);(o,<o'>,<o!>,<o?>,<o/>>,<oe>);\
+	p;q;r;(s,<s<>,<ss>);t;(u,<u'>,<u!>,<u/>>,<u:>);\
+	v;w;x;(y,<y'>,<y:>);(z,<z<>);<aa>;(<a:>,<ae>);(<o:>,<o//>);\
+	<th>;\
+#
+	\{;<NO>;|;\};~;<.M>;<DG>;<My>;<DT>;\
+# remains
+	<-a>;<-o>
