diff -ruN -x obj -x CVS -x *~ -x *orig src56-orig/lib/libc/locale/localeconv.c src/lib/libc/locale/localeconv.c
--- src56-orig/lib/libc/locale/localeconv.c	Fri Sep 19 13:07:17 2014
+++ src/lib/libc/locale/localeconv.c	Fri Sep 19 14:49:15 2014
@@ -33,7 +33,11 @@
 	ret.currency_symbol	= _CurrentMonetaryLocale->currency_symbol;
 	ret.mon_decimal_point	= _CurrentMonetaryLocale->mon_decimal_point;
 	ret.mon_thousands_sep	= _CurrentMonetaryLocale->mon_thousands_sep;
-	ret.mon_grouping	= _CurrentMonetaryLocale->mon_grouping;
+	// Special case to return "" in POSIX locale as shown in column
+	// localeconv() in table of section LC_MONETARY of
+	// http://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap07.html#tag_07_03
+	ret.mon_grouping	= (_CurrentMonetaryLocale == &_DefaultMonetaryLocale) ?
+		"" : (char *)_CurrentMonetaryLocale->mon_grouping;
 	ret.positive_sign	= _CurrentMonetaryLocale->positive_sign;
 	ret.negative_sign	= _CurrentMonetaryLocale->negative_sign;
 	ret.int_frac_digits	= _CurrentMonetaryLocale->int_frac_digits;
