--- src54-mon/gnu/usr.bin/perl/Makefile.bsd-wrapper	Fri Mar 29 21:12:18 2013
+++ src/gnu/usr.bin/perl/Makefile.bsd-wrapper	Wed Dec  4 22:23:25 2013
@@ -17,7 +17,7 @@
 CFLAGS+= -fno-delete-null-pointer-checks
 .endif
 
-CONFIGURE_ARGS= -dsE -Dopenbsd_distribution=defined -Dccflags='-DNO_LOCALE_NUMERIC -DNO_LOCALE_COLLATE'
+CONFIGURE_ARGS= -dsE -Dopenbsd_distribution=defined -Dccflags=''
 
 .if !exists(${.OBJDIR}/config.over)
 CONFIGURE_ARGS+= -Dmksymlinks
@@ -46,7 +46,7 @@
 
 # Arguments passed to Configure...
 
-CPPFLAGS+= -DPERL_CORE -DNO_LOCALE_NUMERIC -DNO_LOCALE_COLLATE -DPERL_RANDOM_DEVICE=\"/dev/arandom\" -I.
+CPPFLAGS+= -DPERL_CORE -DPERL_RANDOM_DEVICE=\"/dev/arandom\" -I.
 
 .if defined (INSTALL_STRIP) && ${INSTALL_STRIP} == "-s"
 STRIPFLAGS='-s'
