diff -ruN src68-orig/gnu/usr.bin/perl/hints/openbsd.sh src/gnu/usr.bin/perl/hints/openbsd.sh
--- src68-orig/gnu/usr.bin/perl/hints/openbsd.sh	Fri Oct  9 12:39:14 2020
+++ src/gnu/usr.bin/perl/hints/openbsd.sh	Thu Oct 22 10:09:26 2020
@@ -187,12 +187,4 @@
 	;;
 esac
 
-# openbsd has a problem regarding newlocale()
-# https://marc.info/?l=openbsd-bugs&m=155364568608759&w=2
-# which is being fixed.  In the meantime, forbid POSIX 2008 locales
-d_newlocale="$undef"
-
-# OpenBSD's locale support is not that complete yet
-ccflags="-DNO_LOCALE_NUMERIC -DNO_LOCALE_COLLATE $ccflags"
-
 # end
