--- src58-orig/gnu/gcc/gcc/config/openbsd.h	Thu Oct 15 18:36:38 2015
+++ src/gnu/gcc/gcc/config/openbsd.h	Wed Dec  9 09:56:18 2015
@@ -86,6 +86,7 @@
   do						\
     {						\
       builtin_define ("__OpenBSD__");		\
+      builtin_define ("__adJ__");		\
       builtin_define ("__unix__");		\
       builtin_define ("__ANSI_COMPAT");		\
       builtin_assert ("system=unix");		\
@@ -117,6 +118,7 @@
   do						\
     {						\
       builtin_define ("__OpenBSD__");		\
+      builtin_define ("__adJ__");		\
       builtin_define ("__unix__");		\
       builtin_define ("__ANSI_COMPAT");		\
       builtin_assert ("system=unix");		\
diff -ruN src65-orig/gnu/llvm/tools/clang/lib/Basic/Targets/OSTargets.h src/gnu/llvm/tools/clang/lib/Basic/Targets/OSTargets.h
--- src65-orig/gnu/llvm/tools/clang/lib/Basic/Targets/OSTargets.h	Sun Jan 27 11:43:02 2019
+++ src/gnu/llvm/tools/clang/lib/Basic/Targets/OSTargets.h	Mon Jul  8 03:11:42 2019
@@ -382,6 +382,7 @@
     // OpenBSD defines; list based off of gcc output
 
     Builder.defineMacro("__OpenBSD__");
+    Builder.defineMacro("__adJ__");
     DefineStd(Builder, "unix", Opts);
     Builder.defineMacro("__ELF__");
     if (Opts.POSIXThreads)
