diff -ruN src61-orig/lib/libc/string/strerror_r.c src/lib/libc/string/strerror_r.c
--- src61-orig/lib/libc/string/strerror_r.c	Wed Oct 19 11:26:16 2016
+++ src/lib/libc/string/strerror_r.c	Sat Jul 22 20:24:13 2017
@@ -5,6 +5,8 @@
 #include <limits.h>
 #include <signal.h>
 #include <string.h>
+#include <xlocale.h>
+#include "locale/xlocale_private.h"
 
 static size_t
 __digits10(unsigned int num)
@@ -83,7 +85,7 @@
 #define	UPREFIX	"Unknown error: "
 
 int
-strerror_r(int errnum, char *strerrbuf, size_t buflen)
+__strerror_rl(int errnum, char *strerrbuf, size_t buflen, locale_t loc)
 {
 	int ret_errno;
 
@@ -94,7 +96,14 @@
 		errno = ret_errno;
 	return (ret_errno);
 }
+
+int
+strerror_r(int errnum, char *strerrbuf, size_t buflen)
+{
+	return __strerror_rl(errnum, strerrbuf, buflen, __get_locale());
+}
 DEF_WEAK(strerror_r);
+
 
 #define USIGPREFIX "Unknown signal: "
 
diff -ruN src62-orig/lib/libc/string/strerror_l.c src/lib/libc/string/strerror_l.c
--- src62-orig/lib/libc/string/strerror_l.c	Mon Sep  4 22:16:13 2017
+++ src/lib/libc/string/strerror_l.c	Sun Dec 10 07:19:53 2017
@@ -17,9 +17,12 @@
 
 #include <limits.h>
 #include <string.h>
+#include <xlocale.h>
 
 #include "thread_private.h"
 
+extern int __strerror_rl(int, char *, size_t, locale_t);
+
 char *
 strerror_l(int errnum, locale_t locale)
 {
@@ -28,6 +31,6 @@
 	char *p = _THREAD_PRIVATE(strerror_l, sel_buf, NULL);
 
 	return p == NULL ? "no buffer available in strerror_l" :
-	    strerror_r(errnum, p, sizeof(sel_buf)) ?
-	    "strerror_r failure" : p;
+	    __strerror_rl(errnum, p, sizeof(sel_buf), locale) ?
+	    "strerror_rl failure" : p;
 }
