diff -ruN src75-orig/gnu/lib/libcxx/include/c++/v1/__config_site src/gnu/lib/libcxx/include/c++/v1/__config_site
--- src75-orig/gnu/lib/libcxx/include/c++/v1/__config_site	Tue Apr  9 16:42:07 2024
+++ src/gnu/lib/libcxx/include/c++/v1/__config_site	Thu May  2 14:08:04 2024
@@ -9,6 +9,8 @@
 #ifndef _LIBCPP___CONFIG_SITE
 #define _LIBCPP___CONFIG_SITE
 
+#define __adJ__
+
 #define _LIBCPP_ABI_VERSION 1
 #define _LIBCPP_ABI_NAMESPACE __1
 /* #undef _LIBCPP_ABI_FORCE_ITANIUM */
diff -ruN src69-orig/gnu/llvm/libcxx/include/__config src/gnu/llvm/libcxx/include/__config
--- src69-orig/gnu/llvm/libcxx/include/__config	Sat Jan  2 16:18:29 2021
+++ src/gnu/llvm/libcxx/include/__config	Sat Jul 24 13:51:15 2021
@@ -631,7 +631,7 @@
 #    define _LIBCPP_DECLARE_STRONG_ENUM_EPILOG(x)
 #  endif // _LIBCPP_CXX03_LANG
 
-#  if defined(__APPLE__) || defined(__FreeBSD__) || defined(_LIBCPP_MSVCRT_LIKE) || defined(__NetBSD__)
+#  if defined(__APPLE__) || defined(__FreeBSD__) || defined(_LIBCPP_MSVCRT_LIKE) || defined(__NetBSD__) || defined(__adJ__)
 #    define _LIBCPP_LOCALE__L_EXTENSIONS 1
 #  endif
 
@@ -891,6 +891,7 @@
 
 #  if defined(__BIONIC__) || defined(__NuttX__) || defined(__Fuchsia__) || defined(__wasi__) ||                        \
       defined(_LIBCPP_HAS_MUSL_LIBC) || defined(__OpenBSD__)
+// In adJ __DefaultRuneLocale is in libc but hidden
 #    define _LIBCPP_PROVIDES_DEFAULT_RUNE_TABLE
 #  endif
 
diff -ruN src69-orig/gnu/llvm/libcxx/include/__locale src/gnu/llvm/libcxx/include/__locale
--- src69-orig/gnu/llvm/libcxx/include/__locale	Sat Jan  2 16:18:29 2021
+++ src/gnu/llvm/libcxx/include/__locale	Sat Jul 24 13:44:23 2021
@@ -346,10 +346,10 @@
   static const mask __regex_word = 0x4000; // 0x8000 and 0x0100 and 0x00ff are used
 #  define _LIBCPP_CTYPE_MASK_IS_COMPOSITE_PRINT
 #  define _LIBCPP_CTYPE_MASK_IS_COMPOSITE_ALPHA
-#elif defined(__APPLE__) || defined(__FreeBSD__) || defined(__NetBSD__)
+#elif defined(__APPLE__) || defined(__FreeBSD__) || defined(__NetBSD__) || defined(__adJ__)
 #  ifdef __APPLE__
   typedef __uint32_t mask;
-#  elif defined(__FreeBSD__)
+#  elif defined(__FreeBSD__) || defined(__adJ__)
   typedef unsigned long mask;
 #  elif defined(__NetBSD__)
   typedef unsigned short mask;
diff -ruN src78-orig/gnu/llvm/libcxx/include/__locale_dir/locale_base_api.h src/gnu/llvm/libcxx/include/__locale_dir/locale_base_api.h
--- src78-orig/gnu/llvm/libcxx/include/__locale_dir/locale_base_api.h	Thu Aug 21 11:00:05 2025
+++ src/gnu/llvm/libcxx/include/__locale_dir/locale_base_api.h	Fri Jan  9 11:24:06 2026
@@ -19,13 +19,17 @@
 #  include <__locale_dir/locale_base_api/solaris.h>
 #elif defined(_NEWLIB_VERSION)
 #  include <__locale_dir/locale_base_api/newlib.h>
-#elif defined(__OpenBSD__)
+#elif defined(__OpenBSD__) && !defined(__adJ__)
 #  include <__locale_dir/locale_base_api/openbsd.h>
 #elif defined(__Fuchsia__)
 #  include <__locale_dir/locale_base_api/fuchsia.h>
 #elif defined(__wasi__) || defined(_LIBCPP_HAS_MUSL_LIBC)
 #  include <__locale_dir/locale_base_api/musl.h>
 #elif defined(__APPLE__) || defined(__FreeBSD__)
+#  include <xlocale.h>
+#elif defined(__adJ__)
+#  include <stdlib.h>
+#  include <stdio.h>
 #  include <xlocale.h>
 #endif
 
