diff -ruN src72-orig/etc/daily src/etc/daily
--- src72-orig/etc/daily	Tue Oct 20 17:42:29 2020
+++ src/etc/daily	Tue Apr 11 21:38:25 2023
@@ -50,7 +50,7 @@
 	find -x . \
 	    \( -path './ssh-*' -o -path ./.X11-unix -o -path ./.ICE-unix \
 		-o -path './tmux-*' \) \
-	    -prune -o -type f -atime +7 -delete 2>/dev/null
+	    -prune -o -type f -atime +7 -a ! -name "*shm" -delete 2>/dev/null
 	find -x . -type d -mtime +1 ! -path ./vi.recover ! -path ./.X11-unix \
 	    ! -path ./.ICE-unix ! -name . \
 	    -delete >/dev/null 2>&1; }

